---
title: "sleep_GWAS"
author: "Claire Morrison"
date: "6/13/2022"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(data.table)
library(dplyr)
library(ggplot2)
library(ggpubr)
library(stringr)
library(wordcloud)
```


02/10/2022: running GWAS on sleep health factors in genomic SEM

This script breaks down the genomic SEM GWAS process into steps made easy with array jobs on a super computer. There is a combination of bash and R scripts, but the big workhorse is the bash script that calls the R GWAS script in Step 5.

The directory set up should be:

~code/ --> where the code for each array job goes
~results/F1 --> results for each factor
~results/F2
~results/F3
~results/F4
~results/F5
~results/F6
~outerr --> error and output files


I will split the sumstats into about 7,000 files (10,000 SNPs per file), and then run th GWAS on each file. So if there's 7,000 sumstsats files, there will be 7,000 array jobs, 7,000 .R files in the ~/code directory, 7,000 results files etc.



########################################################
#### trying again with new sleep structure after R&R

Step 1: Get sumstats files in Genomic SEM

```{r, eval=F}

setwd("pl/active/friedman/studies/claire_ldsc_gsem/ldsc/sleepsumstats/cleaned")

require(GenomicSEM)
files<-c("chronotype.txt", "midR.txt", "l5R.txt", "m10R.txt",
         "episodesR.txt", "effic.txt",
        "sleepinessR.txt", "napsR.txt", "diurnR.txt",
         "selfsleepdur.txt",
         "sdsleepdurR.txt",
        "insomnia_withNeffR.txt")

#2. ref = the name of the reference file used to obtain SNP MAF
ref="../../../gsem/reference.1000G.maf.0.005.txt"

#3. trait.names = the name of the files to be used in 
trait.names=c("chron","midR", "l5R", "m10R",
               "episodesR", "efficiency",
               "sleepinessR", "nappingR", "diurnalR",
               "self_sleepdur",
               "sd_sleepdurR",
               "insomR")

#4. se.logit = whether the standard errors are on an logistic scale
se.logit<-c(T,F,F,F,F,F,F,F,F,F,F,F)
OLS<- c(F,T,T,T,T,T,T,T,T,T,T,T)
#run the sumstats function below
sleep_sumstats <- sumstats(files=files,ref=ref ,trait.names=trait.names,se.logit=se.logit)

save(sleep_sumstats, file="../../sleep_sumstats.RData")
```

Sleep sumstats were created, and the outfile was renamed to sleep_sumstats_out.txt:

"After merging across all summary statistics using listwise deletion, performing QC, and merging with the reference file, there are 7188521 SNPs left in the final multivariate summary statistics file"


Step 3: Run covstruct

```{r, eval=F}
#vector of munged summary statisitcs
traits<-c("munged/chrono.sumstats.gz", "munged/midR.sumstats.gz", "munged/l5R.sumstats.gz", "munged/m10R.sumstats.gz",
         "munged/episodesR.sumstats.gz", "munged/effic.sumstats.gz",
         "munged/sleepinessR.sumstats.gz", "munged/napsR.sumstats.gz", "munged/diurnR.sumstats.gz",
         "munged/selfsleepdur.sumstats.gz",
         "munged/sdsleepdurR.sumstats.gz",
         "munged/insomR.sumstats.gz")

#enter sample prevalence of .5 to reflect that all traits were munged using the sum of effective sample size
sample.prev<-c(NA, NA, NA, NA, 
               NA, NA, 
               NA, NA, NA,
               NA,
               NA,
               .54)

#vector of population prevalences
population.prev<-c(NA, NA, NA, NA, 
               NA, NA, 
               NA, NA, NA,
               NA,
               NA,
               .30)

#the folder of LD scores
ld<-"../eur_w_ld_chr/"

#the folder of LD weights [typically the same as folder of LD scores]
wld<-"../eur_w_ld_chr/"

#name the traits
trait.names<-c("chron","midR", "l5R", "m10R",
               "episodesR", "efficiency",
               "sleepinessR", "nappingR", "diurnalR",
               "self_sleepdur",
               "sd_sleepdurR",
               "insomR")

#run LDSC
sleepLDSC<-ldsc(traits=traits,sample.prev=sample.prev,population.prev=population.prev,ld=ld,wld=wld,trait.names=trait.names)

#optional command to save the output as a .RData file for later use
save(sleepLDSC,file="../../sleepLDSC_GWAS_06_2022.RData")#vector of munged summary statisitcs

```


Step 3: Split sumstats

/pl/active/friedman/studies/claire_ldsc_gsem/gsem/sleep_sumstats_06_2022.RData = sumstats


```{r}
### load summary statistics RData file generated by genomicSEM
load("sleep_sumstats_06_2022.RData")

### split the summary statistics SNPs into smaller sets so that they can be processed faster
### save the new list of dataframes as .RData file
split_sumstats <- split(sleep_sumstats, (as.numeric(rownames(sleep_sumstats))-1) %/% 1000)
save(split_sumstats, file="split_sumstats_06_2022.RData")

### determine number of SNP sets and then save to a file
num_SNP_sets <- length(split_sumstats)
write.table(num_SNP_sets, file="/scratch/summit/clmo1617/gsem.num_SNP_sets.txt", row.names=FALSE, col.names=FALSE, quote=FALSE)

```

The sumstats have been split into 7413 sets. 

Step 4: Save and identify model

This chunk is saved as sleepfactormodel_06_2022.R

/pl/active/friedman/studies/claire_ldsc_gsem/gsem/sleepLDSC_GWAS_06_2022.RData = LDSC
/pl/active/friedman/studies/claire_ldsc_gsem/gsem/split_sumstats_06_2022.RData = split sumstats

```{r sleep GWAS}


### load necessary packages
library(devtools)
require(GenomicSEM)
library(dplyr)


load("/pl/active/friedman/studies/claire_ldsc_gsem/gsem/split_sumstats_06_2022.RData")
load("/pl/active/friedman/studies/claire_ldsc_gsem/gsem/sleepLDSC_GWAS_06_2022.RData")

model<-"
F1=~efficiency+episodesR
F2=~sleepinessR+diurnalR+nappingR
F3=~chron+midR+m10R+l5R
F4=~self_sleepdur
F5=~sd_sleepdurR
F6=~insomR

sd_sleepdurR~~0*sd_sleepdurR
insomR~~0*insomR
self_sleepdur~~0*self_sleepdur

efficiency~~a*efficiency
a>.001

F1~SNP
F2~SNP
F3~SNP
F4~SNP
F5~SNP
F6~SNP
"

### run a user model for one subset of SNPs

### covstruct is created ahead of time
### split sumstats is created ahead of time
CommonFactor <- userGWAS(covstruc=sleepLDSC, SNPs=split_sumstats[[NUMBER]], model=model, sub=c("F1~SNP", "F2~SNP", "F3~SNP"),cores=3, parallel=T)

CommonFactor <- userGWAS(covstruc=sleepLDSC, SNPs=test, model=model, sub=c("F1~SNP", "F2~SNP", "F3~SNP"),cores=3, parallel=F)


### clean & save sumstats files
factors<- c(1,2,3)

for (i in factors) {
  CommonFactor[[i]]$free <- NULL
  CommonFactor[[i]]$label <- NULL
  CommonFactor[[i]]$lhs <- NULL
  CommonFactor[[i]]$op <- NULL
  CommonFactor[[i]]$rhs <- NULL
  
  DF <- filter(CommonFactor[[i]], warning==0)
  
  write.table(DF, file=paste0("./results/F",i,"/NUMBER.txt"), row.names=FALSE, quote=FALSE) 
}
```


Run GWAS in batch jobs

```{r}
#SBATCH -J multi-GWAS
#SBATCH --partition=amilan
#SBATCH --ntasks=3
#SBATCH --output=./outerr/test.%a.out
#SBATCH --error=./outerr/test.%a.err
#SBATCH --mem=70gb
#SBATCH --array=1-10
#SBATCH --time=2:00:00

export OMP_NUM_THREADS=1

### display time at beginning and end of script to track how long it takes 
date
### display node that job was run on
hostname

### set variable that will be the number of the SNP set
cc="${SLURM_ARRAY_TASK_ID}"

### replace "NUMBER" variable in R script with the number of the SNP set the job is running
### you can delete these scripts after you're done with the analyses
sed "s/NUMBER/$cc/g" 2.sleepGSEM.R > ./code/$cc.R
ml slurm/alpine

### run the Rscript
ml load R
Rscript ./code/$cc.R

date

```



Use this to get the jobs that did not finish -- paste into array job
```{r}
## had some smaller jobs that were not finishing
find ./ -type f -size -44k | xargs rm -f


ml R
R

library(stringr)
completed<- as.data.frame(list.files())
colnames(completed)[1]<- 'l1'
completed$V1<- str_split(completed$l1,"[.]", simplify=T)[,1]
final<- seq(1,7413,1)
diff<- setdiff(final, completed$V1)
write.table(matrix(as.character(diff),nrow=1), "/scratch/summit/clmo1617/gsem/sleepredos.csv",sep=",", row.names=FALSE, col.names=FALSE, quote = F)


### might need to paste that output into a text file to make sure it's all one line, then paste as array jobs
```

concatenate all files

```{r}
#!/bin/bash

#SBATCH --job-name=combine_SNPsets
#SBATCH --time=02:00:00
#SBATCH --partition=shas
#SBATCH -n 1
#SBATCH --mem=1000
#SBATCH --output=./outerr/combine.out
#SBATCH --error=./outerr/combine.err
#SBATCH --array=1-3

date
hostname

### create variable that is equal to the number of factors (3)
cc="${SLURM_ARRAY_TASK_ID}"

### concatenate all results files
cat results/F"$cc"/*.txt > F"$cc".txt

rm ./code/*.R
### remove all lines that aren't the first occurrence of Z_Estimate (used to indicate header line)
sed -i '1!{/Z_Estimate/d;}' F"$cc".txt


date

```



### Step 7: calc eff N for FUMA/MAGMA
rm(list = ls(all.names = TRUE))

Do this then divide estimate by scales derived below. Used the wrong identification so standard errors are larger (while p values etc are the same). Ran the same 10,000 SNPs with both types of identificaiton and computed the scales of the effective N estiamtes.


Efficiency scaled by 9.000123 = 31,654.47
Alertness scaled by 13.40819 =  1,043,563
Chronotype scaled by 11.14108 = 79,508.29


```{r}
### alert N hat = 13992293
f2<- fread("F2_sumstats.txt", header=T, data.table=F)
#restrict to MAF of 40% and 10%
alert_factor<-subset(f2, f2$MAF <= .4 & f2$MAF >= .1)
#calculate expected sample size (N_hat)
N_hat<-mean(1/((2*alert_factor$MAF*(1-alert_factor$MAF))*alert_factor$SE^2))
alert_factor$N_eff<- rep(13992293,1,nrow(alert_factor))

f2<- fread("F2_sumstats.txt", header=T, data.table=F)
f2$N_eff<- rep(13992293,1,nrow(f2))
fwrite(f2, "/pl/active/friedman/studies/claire_ldsc_gsem/sleep_rr_upd/alertness_gsem.txt", sep='\t')

### effic N-hat = 284894.1
f1<- fread("F1_sumstats.txt", header=T, data.table=F)
#restrict to MAF of 40% and 10%
effic_factor<-subset(f1, f1$MAF <= .4 & f1$MAF >= .1)
#calculate expected sample size (N_hat)
N_hat<-mean(1/((2*effic_factor$MAF*(1-effic_factor$MAF))*effic_factor$SE^2))


f1<- fread("F1_sumstats.txt", header=T, data.table=F)
f1$N_eff<- rep(284894.1,1,nrow(f1))
fwrite(f1, "/pl/active/friedman/studies/claire_ldsc_gsem/sleep_rr_upd/efficiency_gsem.txt", sep='\t')

### chrono N hat = 885808.2
f3<- fread("F3_sumstats.txt", header=T, data.table=F)
#restrict to MAF of 40% and 10%
chrono_factor<-subset(f3, f3$MAF <= .4 & f3$MAF >= .1)
#calculate expected sample size (N_hat)
N_hat<-mean(1/((2*chrono_factor$MAF*(1-chrono_factor$MAF))*chrono_factor$SE^2))


f3<- fread("F3_sumstats.txt", header=T, data.table=F)
f3$N_eff<- rep(885808.2,1,nrow(f3))
fwrite(f3, "/pl/active/friedman/studies/claire_ldsc_gsem/sleep_rr_upd/circadian_pref_gsem.txt", sep='\t')

```




```{r}
############ testing


### per Andrew: effective N will b huge if you use variance ID. assuming that is the default? I tried default vs unit loading ID vs variance ID (below) and the p values, fit etc are all the same -- which is reassuring

# default

library(devtools)
require(GenomicSEM)
library(dplyr)


load("/pl/active/friedman/studies/claire_ldsc_gsem/gsem/split_sumstats_06_2022.RData")
load("/pl/active/friedman/studies/claire_ldsc_gsem/gsem/sleepLDSC_GWAS_06_2022.RData")

test<- split_sumstats[[5]][1:10,]

model<-"
F1=~efficiency+episodesR
F2=~sleepinessR+diurnalR+nappingR
F3=~chron+midR+m10R+l5R
F4=~self_sleepdur
F5=~sd_sleepdurR
F6=~insomR

sd_sleepdurR~~0*sd_sleepdurR
insomR~~0*insomR
self_sleepdur~~0*self_sleepdur

efficiency~~a*efficiency
a>.001

F1~SNP
F2~SNP
F3~SNP
F4~SNP
F5~SNP
F6~SNP
"


CommonFactor <- userGWAS(covstruc=sleepLDSC, SNPs=test, model=model, sub=c("F1~SNP", "F2~SNP", "F3~SNP"),cores=3, parallel=T)
fwrite(CommonFactor[[1]], "effic_default_gwas.txt", sep='\t') 
fwrite(CommonFactor[[3]], "chron_default_gwas.txt", sep='\t') 

# N eff 1
default<- fread("effic_default_gwas.txt", header=T, data.table=F)
#restrict to MAF of 40% and 10%
default_sub<-subset(default, default$MAF <= .4 & default$MAF >= .1)
#calculate expected sample size (N_hat)
default_N_hat<-mean(1/((2*default_sub$MAF*(1-default_sub$MAF))*default_sub$SE^2))

colnames(default)[2:ncol(default)]<-paste(colnames(default)[2:ncol(default)],"default",sep="_")


# unit ID

model<-"
F1=~1*efficiency+episodesR
F2=~1*nappingR+sleepinessR+diurnalR
F3=~1*chron+midR+m10R+l5R
F4=~1*self_sleepdur
F5=~1*sd_sleepdurR
F6=~1*insomR

sd_sleepdurR~~0*sd_sleepdurR
insomR~~0*insomR
self_sleepdur~~0*self_sleepdur

efficiency~~a*efficiency
a>.001

F1~SNP
F2~SNP
F3~SNP
F4~SNP
F5~SNP
F6~SNP
"

unitID <- userGWAS(covstruc=sleepLDSC, SNPs=test, model=model, sub=c("F1~SNP", "F2~SNP", "F3~SNP"),cores=3, parallel=T)
fwrite(unitID[[1]], "effic_unitID_gwas.txt", sep='\t') 
fwrite(unitID[[3]], "chron_unitID_gwas.txt", sep='\t') 

# N eff 1
unitID<- fread("effic_unitID_gwas.txt", header=T, data.table=F)
#restrict to MAF of 40% and 10%
unitID_sub<-subset(unitID, unitID$MAF <= .4 & unitID$MAF >= .1)
#calculate expected sample size (N_hat)
unitID_N_hat<-mean(1/((2*unitID_sub$MAF*(1-unitID_sub$MAF))*unitID_sub$SE^2))

colnames(unitID)[2:ncol(unitID)]<-paste(colnames(unitID)[2:ncol(unitID)],"unitID",sep="_")



# var ID

model<-"
F1=~NA*efficiency+episodesR
F2=~NA*nappingR+sleepinessR+diurnalR
F3=~NA*chron+midR+m10R+l5R
F4=~1*self_sleepdur
F5=~1*sd_sleepdurR
F6=~1*insomR

sd_sleepdurR~~0*sd_sleepdurR
insomR~~0*insomR
self_sleepdur~~0*self_sleepdur

efficiency~~a*efficiency
a>.001

F1~~1*F1
F2~~1*F2
F3~~1*F3


F1~SNP
F2~SNP
F3~SNP
F4~SNP
F5~SNP
F6~SNP
"

varID <- userGWAS(covstruc=sleepLDSC, SNPs=test, model=model, sub=c("F1~SNP", "F2~SNP", "F3~SNP"),cores=3, parallel=T)
fwrite(varID[[1]], "effic_varID_gwas.txt", sep='\t') 
fwrite(varID[[3]], "chron_varID_gwas.txt", sep='\t') 

# N eff 1
varID<- fread("effic_varID_gwas.txt", header=T, data.table=F)
#restrict to MAF of 40% and 10%
varID_sub<-subset(varID, varID$MAF <= .4 & varID$MAF >= .1)
#calculate expected sample size (N_hat)
varID_N_hat<-mean(1/((2*varID_sub$MAF*(1-varID_sub$MAF))*varID_sub$SE^2))

colnames(varID)[2:ncol(varID)]<-paste(colnames(varID)[2:ncol(varID)],"VarID",sep="_")



# merge
d<- merge(default, unitID, by='SNP')
d2<- merge(d, varID, by='SNP')

# unit and default are the same
cor.test(d2$est_unitID,d2$est_default)
cor.test(d2$Pval_Estimate_unitID,d2$Pval_Estimate_default)

# unit and var
cor.test(d2$Pval_Estimate_unitID,d2$Pval_Estimate_VarID)
cor.test(d2$est_unitID,d2$est_VarID)
cor.test(d2$chisq_unitID,d2$chisq_VarID)

unitID_N_hat - varID_N_hat = 241569.8


```

try to figure out scale of effective N
GWAS on ~ 10,000 SNPs both ways
```{r unit var}
library(devtools)
require(GenomicSEM)
library(dplyr)


load("/pl/active/friedman/studies/claire_ldsc_gsem/gsem/split_sumstats_06_2022.RData")
load("/pl/active/friedman/studies/claire_ldsc_gsem/gsem/sleepLDSC_GWAS_06_2022.RData")

model<-"
F1=~NA*efficiency+episodesR
F2=~NA*nappingR+sleepinessR+diurnalR
F3=~NA*chron+midR+m10R+l5R
F4=~1*self_sleepdur
F5=~1*sd_sleepdurR
F6=~1*insomR

sd_sleepdurR~~0*sd_sleepdurR
insomR~~0*insomR
self_sleepdur~~0*self_sleepdur

efficiency~~a*efficiency
a>.001

F1~~1*F1
F2~~1*F2
F3~~1*F3


F1~SNP
F2~SNP
F3~SNP
F4~SNP
F5~SNP
F6~SNP
"
CommonFactor <- userGWAS(covstruc=sleepLDSC, SNPs=split_sumstats[[NUMBER]], model=model, sub=c("F1~SNP", "F2~SNP", "F3~SNP"),cores=3, parallel=T)


### clean & save sumstats files
factors<- c(1,2,3)

for (i in factors) {
  CommonFactor[[i]]$free <- NULL
  CommonFactor[[i]]$label <- NULL
  CommonFactor[[i]]$lhs <- NULL
  CommonFactor[[i]]$op <- NULL
  CommonFactor[[i]]$rhs <- NULL
  
  DF <- filter(CommonFactor[[i]], warning==0)
  
  write.table(DF, file=paste0("./results_unitvar/F",i,"/NUMBER.txt"), row.names=FALSE, quote=FALSE) 
}
```

```{r unit load}
library(devtools)
require(GenomicSEM)
library(dplyr)


load("/pl/active/friedman/studies/claire_ldsc_gsem/gsem/split_sumstats_06_2022.RData")
load("/pl/active/friedman/studies/claire_ldsc_gsem/gsem/sleepLDSC_GWAS_06_2022.RData")


model<-"
F1=~1*efficiency+episodesR
F2=~1*nappingR+sleepinessR+diurnalR
F3=~1*chron+midR+m10R+l5R
F4=~1*self_sleepdur
F5=~1*sd_sleepdurR
F6=~1*insomR

sd_sleepdurR~~0*sd_sleepdurR
insomR~~0*insomR
self_sleepdur~~0*self_sleepdur

efficiency~~a*efficiency
a>.001

F1~SNP
F2~SNP
F3~SNP
F4~SNP
F5~SNP
F6~SNP
"

unitID <- userGWAS(covstruc=sleepLDSC, SNPs=test, model=model, sub=c("F1~SNP", "F2~SNP", "F3~SNP"),cores=3, parallel=T)

CommonFactor <- userGWAS(covstruc=sleepLDSC, SNPs=split_sumstats[[NUMBER]], model=model, sub=c("F1~SNP", "F2~SNP", "F3~SNP"),cores=3, parallel=T)


### clean & save sumstats files
factors<- c(1,2,3)

for (i in factors) {
  CommonFactor[[i]]$free <- NULL
  CommonFactor[[i]]$label <- NULL
  CommonFactor[[i]]$lhs <- NULL
  CommonFactor[[i]]$op <- NULL
  CommonFactor[[i]]$rhs <- NULL
  
  DF <- filter(CommonFactor[[i]], warning==0)
  
  write.table(DF, file=paste0("./results_unitload/F",i,"/NUMBER.txt"), row.names=FALSE, quote=FALSE) 
}
```

Effective N scales
#Efficiency scaled by 9.000123 --- 284894.1 = 31,654.47
#Alertness scaled by 13.40819 --- 13992293 = 1,043,563
#Chronotype scaled by 11.14108 ---- 885808.2 = 79,508.29
```{r eff N scales}
# effic
loadID<- fread("F1_sumstats_unitload.txt", header=T, data.table=F)
varID<- fread("F1_sumstats_unitvar.txt", header=T, data.table=F)

varID_sub<-subset(varID, varID$MAF <= .4 & varID$MAF >= .1)
varID_N_hat<-mean(1/((2*varID_sub$MAF*(1-varID_sub$MAF))*varID_sub$SE^2))

loadID_sub<-subset(loadID, loadID$MAF <= .4 & loadID$MAF >= .1)
loadID_N_hat<-mean(1/((2*loadID_sub$MAF*(1-loadID_sub$MAF))*loadID_sub$SE^2))

loadID_N_hat/varID_N_hat #Efficiency scaled by 9.000123

rm(list = ls(all.names = TRUE))

# alert
loadID<- fread("F2_sumstats_unitload.txt", header=T, data.table=F)
varID<- fread("F2_sumstats_unitvar.txt", header=T, data.table=F)

varID_sub<-subset(varID, varID$MAF <= .4 & varID$MAF >= .1)
varID_N_hat<-mean(1/((2*varID_sub$MAF*(1-varID_sub$MAF))*varID_sub$SE^2))

loadID_sub<-subset(loadID, loadID$MAF <= .4 & loadID$MAF >= .1)
loadID_N_hat<-mean(1/((2*loadID_sub$MAF*(1-loadID_sub$MAF))*loadID_sub$SE^2))

loadID_N_hat/varID_N_hat #Alertness scaled by 13.40819

rm(list = ls(all.names = TRUE))

# chrono
loadID<- fread("F3_sumstats_unitload.txt", header=T, data.table=F)
varID<- fread("F3_sumstats_unitvar.txt", header=T, data.table=F)

varID_sub<-subset(varID, varID$MAF <= .4 & varID$MAF >= .1)
varID_N_hat<-mean(1/((2*varID_sub$MAF*(1-varID_sub$MAF))*varID_sub$SE^2))

loadID_sub<-subset(loadID, loadID$MAF <= .4 & loadID$MAF >= .1)
loadID_N_hat<-mean(1/((2*loadID_sub$MAF*(1-loadID_sub$MAF))*loadID_sub$SE^2))

loadID_N_hat/varID_N_hat #Chronotype scaled by 11.14108


```

Look at variation in standard errors
```{r}
# effic
loadID<- fread("F1_sumstats_unitload.txt", header=T, data.table=F)
colnames(loadID)[2:ncol(loadID)]<-paste(colnames(loadID)[2:ncol(loadID)],"loadID",sep="_")


varID<- fread("F1_sumstats_unitvar.txt", header=T, data.table=F)
colnames(varID)[2:ncol(varID)]<-paste(colnames(varID)[2:ncol(varID)],"varID",sep="_")

effic<- merge(loadID, varID, by='SNP')
cor.test(effic$SE_loadID, effic$SE_varID) #0.999888

effic$se_ratio<- effic$SE_loadID/effic$SE_varID

fwrite(effic, "effictestID.txt", sep='\t')

rm(list = ls(all.names = TRUE))

# alert
loadID<- fread("F2_sumstats_unitload.txt", header=T, data.table=F)
colnames(loadID)[2:ncol(loadID)]<-paste(colnames(loadID)[2:ncol(loadID)],"loadID",sep="_")


varID<- fread("F2_sumstats_unitvar.txt", header=T, data.table=F)
colnames(varID)[2:ncol(varID)]<-paste(colnames(varID)[2:ncol(varID)],"varID",sep="_")

alert<- merge(loadID, varID, by='SNP')
cor.test(alert$SE_loadID, alert$SE_varID) #0.9999863

alert$se_ratio<- alert$SE_loadID/alert$SE_varID

fwrite(alert, "alerttestID.txt", sep='\t')

rm(list = ls(all.names = TRUE))

# chrono

loadID<- fread("F3_sumstats_unitload.txt", header=T, data.table=F)
colnames(loadID)[2:ncol(loadID)]<-paste(colnames(loadID)[2:ncol(loadID)],"loadID",sep="_")


varID<- fread("F3_sumstats_unitvar.txt", header=T, data.table=F)
colnames(varID)[2:ncol(varID)]<-paste(colnames(varID)[2:ncol(varID)],"varID",sep="_")

chrono<- merge(loadID, varID, by='SNP')
cor.test(chrono$SE_loadID, chrono$SE_varID) #0.9999947

chrono$se_ratio<- chrono$SE_loadID/chrono$SE_varID

fwrite(chrono, "chronotestID.txt", sep='\t')
```

plot 
```{r}
library(data.table)
library(ggplot2)
library(dplyr)
effic<-fread("/Users/clairemorrison/Desktop/IDtest/effictestID.txt", header=T, data.table=F)
effic<- effic %>% mutate(trait='Efficiency')%>%
  select(trait, se_ratio) 

alert<-fread("/Users/clairemorrison/Desktop/IDtest/alerttestID.txt", header=T, data.table=F)
alert<- alert %>% mutate(trait='Alertness')%>%
  select(trait, se_ratio)


chrono<-fread("/Users/clairemorrison/Desktop/IDtest/chronotestID.txt", header=T, data.table=F)
chrono<- chrono %>% mutate(trait='Circadian Preference')%>%
  select(trait, se_ratio)

dat<- rbind(effic,alert,chrono)

ggplot(dat, aes(x=se_ratio, fill=trait)) + geom_density(alpha=.3) + 
  labs(title = "Unit loading identification SEs divided by unit variance identification SEs for GWAS factors", x = "Ratio of standard errors", y = "Density")

```

Make sure GWAS was run with factor correlations

```{r}
load("/pl/active/friedman/studies/claire_ldsc_gsem/gsem/split_sumstats_06_2022.RData")
load("/pl/active/friedman/studies/claire_ldsc_gsem/gsem/sleepLDSC_GWAS_06_2022.RData")

test<- split_sumstats[[5]][1:10,]

#### default
model<-"
F1=~efficiency+episodesR
F2=~sleepinessR+diurnalR+nappingR
F3=~chron+midR+m10R+l5R
F4=~self_sleepdur
F5=~sd_sleepdurR
F6=~insomR

sd_sleepdurR~~0*sd_sleepdurR
insomR~~0*insomR
self_sleepdur~~0*self_sleepdur

efficiency~~a*efficiency
a>.001

F1~SNP
F2~SNP
F3~SNP
F4~SNP
F5~SNP
F6~SNP
"

CommonFactor <- userGWAS(covstruc=sleepLDSC, SNPs=test, model=model, sub=c("F1~SNP", "F2~SNP", "F3~SNP"),cores=3, parallel=T)

### compare output to
model<-"
F1=~efficiency+episodesR
F2=~sleepinessR+diurnalR+nappingR
F3=~chron+midR+m10R+l5R
F4=~self_sleepdur
F5=~sd_sleepdurR
F6=~insomR

sd_sleepdurR~~0*sd_sleepdurR
insomR~~0*insomR
self_sleepdur~~0*self_sleepdur

efficiency~~a*efficiency
a>.001

F1~~F2+F3+F4+F5+F6
F2~~F3+F4+F5+F6
F3~~F4+F5+F6
F4~~F5+F6
F5~~F6

F1~SNP
F2~SNP
F3~SNP
F4~SNP
F5~SNP
F6~SNP
"

CommonFactor_withcorrs <- userGWAS(covstruc=sleepLDSC, SNPs=test, model=model, sub=c("F1~SNP", "F2~SNP", "F3~SNP"),cores=3, parallel=T)


### same DF and ests etc
```

### Q SNP analyses

```{r}

library(devtools)
require(GenomicSEM)
library(dplyr)


load("/pl/active/friedman/studies/claire_ldsc_gsem/gsem/split_sumstats_06_2022.RData")
load("/pl/active/friedman/studies/claire_ldsc_gsem/gsem/sleepLDSC_GWAS_06_2022.RData")


model<-"
F1=~efficiency+episodesR
F2=~sleepinessR+diurnalR+nappingR
F3=~chron+midR+m10R+l5R
F4=~self_sleepdur
F5=~sd_sleepdurR
F6=~insomR

sd_sleepdurR~~0*sd_sleepdurR
insomR~~0*insomR
self_sleepdur~~0*self_sleepdur

efficiency~~a*efficiency
a>.001


efficiency+episodesR+sleepinessR+diurnalR+nappingR+chron+midR+m10R+l5R+self_sleepdur+sd_sleepdurR+insomR~SNP
"

CommonFactor <- userGWAS(covstruc=sleepLDSC, SNPs=split_sumstats[[NUMBER]], model=model, sub=c("efficiency~SNP", "episodesR~SNP", "sleepinessR~SNP", "diurnalR~SNP", "nappingR~SNP", "chron~SNP", "midR~SNP", "m10R~SNP", "l5R~SNP"), parallel=T)


### clean & save sumstats files
factors<- seq(1:9)

for (i in factors) {
  CommonFactor[[i]]$free <- NULL
  CommonFactor[[i]]$label <- NULL
  CommonFactor[[i]]$lhs <- NULL
  CommonFactor[[i]]$op <- NULL
  CommonFactor[[i]]$rhs <- NULL
  
  DF <- filter(CommonFactor[[i]], warning==0)
  
  write.table(DF, file=paste0("./results/indicator",i,"/NUMBER.txt"), row.names=FALSE, quote=FALSE) 
}
```


```{r}

ml anaconda
conda activate /projects/lessem/software/anaconda/envs/R-4.2.1

R

library(stringr)
completed<- as.data.frame(list.files())
colnames(completed)[1]<- 'l1'
completed$V1<- str_split(completed$l1,"[.]", simplify=T)[,1]
final<- seq(1,7413,1)
diff<- setdiff(final, completed$V1)
write.table(matrix(as.character(diff),nrow=1), "/scratch/alpine/clmo1617/gsem/sleepredos2.csv",sep=",", row.names=FALSE, col.names=FALSE, quote = F)




```

Concatenate (make sure you save chi square!!!)

```{r}
#!/bin/bash

#SBATCH --job-name=combine_SNPsets
#SBATCH --time=02:00:00
#SBATCH --partition=amilan
#SBATCH -n 1
#SBATCH --mem=1000
#SBATCH --output=./outerr/combine.out
#SBATCH --error=./outerr/combine.err
#SBATCH --array=1-9

date
hostname

### create variable that is equal to the number of factors (3)
cc="${SLURM_ARRAY_TASK_ID}"

### concatenate all results files
cat results/indicator"$cc"/*.txt > indicator"$cc"_Q.txt
### remove error and warning columnsc
#awk '{$11=$12=""; print $0}' indicator"$cc"_stats.txt > indicator"$cc"_Q.txt
### remove intermediate files
#rm indicator"$cc"_stats.txt
rm ./code/*.R
### remove all lines that aren't the first occurrence of Z_Estimate (used to indicate header line)
sed -i '1!{/Z_Estimate/d;}' indicator"$cc"_Q.txt


date
```



# calc chi square and p values 

F1,F2 and F3 are just the sum stats from each factor (F1= effic, F2= alert and F3= circ)
They are named this because I removed the chi square value in the sumstats I saved
Merge F1 with really any indicator and compute chi square diff test. 
Then get indicator~SNP beta and p for each indicator and merge into one file

**FIRST** just compute chi square difference test with factor 1 and any indicator -- this should be the same across all factors and indicators

```{r}


ml anaconda
conda activate /projects/lessem/software/anaconda/envs/R-4.2.1

R

library(data.table)
library(dplyr)

# read in files
f1<- fread("F1_Q.txt", header=T, data.table=F)
i1<- fread("indicator1.txt", header=T, data.table=F)

# change column names so can merge
colnames(f1)[7:ncol(f1)]<- paste("eff_factor", colnames(f1)[2:ncol(f1)], sep="_")
colnames(i1)[2:10]<- paste("efficiency", colnames(i1)[2:10], sep="_")

f1<- f1 %>% select(SNP, CHR, BP, MAF, A1, A2, eff_factor_est, eff_factor_SE, eff_factor_Pval_Estimate)
i1<- i1 %>% select(SNP, chisq, chisq_df, chisq_pval, efficiency_est, efficiency_SE, efficiency_Pval_Estimate)

# merge 
Qtest<- merge(f1,i1,by='SNP')

# calc chi square difference test
Qtest$chisq<- as.numeric(Qtest$chisq)
Qtest$chisq_df<- as.numeric(Qtest$chisq_df)

Qtest$Q_chisq<-Qtest$eff_factor_chisq-Qtest$chisq
Qtest$df<-Qtest$eff_factor_chisq_df-Qtest$chisq_df
Qtest$Q_chisq_pval<-pchisq(Qtest$Q_chisq,Qtest$df,lower.tail=FALSE)

sigQ<- subset(Qtest, Qtest$Q_chisq_pval<5e-8) ### 1332 significant chi square difference tests

fwrite(sigQ, "Q_sig.txt", sep='\t')

rm(list = ls(all.names = TRUE))
```


Efficiency

now -- merge in efficiency indicators to the subsetted file
```{r}

sigQ<- fread("Q_sig.txt", header=T, data.table=F)

i2<- fread("indicator2.txt", header=T, data.table=F)

colnames(i2)[2:10]<- paste("episodes", colnames(i2)[2:10], sep="_")

# select only relevant columns
# only need chi square, df and p for one indicator (same for all)
i2<- i2 %>% select(SNP, episodes_est, episodes_SE, episodes_Pval_Estimate)

# merge
sigQ<- merge(sigQ,i2, by='SNP')


sigQ$sigEff<- ifelse(sigQ$efficiency_Pval_Estimate<5e-8,1,0)
sigQ$sigEp<- ifelse(sigQ$episodes_Pval_Estimate<5e-8,1,0)

# 2 sig Q SNPs for efficiency indicator: rs182588061 & rs6735071

```

now merge in alertness indicators *and* factor so everything is in one place
```{r}
rm(list = ls(all.names = TRUE))

sigQ<- fread("Q_sig.txt", header=T, data.table=F)

f2<- fread("F2_Q.txt", header=T, data.table=F)
i3<- fread("indicator3.txt", header=T, data.table=F)
i4<- fread("indicator4.txt", header=T, data.table=F)
i5<- fread("indicator5.txt", header=T, data.table=F)

colnames(f2)[2:ncol(f2)]<- paste("alert_factor", colnames(f2)[2:ncol(f2)], sep="_")
colnames(i3)[2:10]<- paste("sleepiness", colnames(i3)[2:10], sep="_")
colnames(i4)[2:10]<- paste("diurnal", colnames(i4)[2:10], sep="_")
colnames(i5)[2:10]<- paste("napping", colnames(i5)[2:10], sep="_")

f2<- f2 %>% select(SNP, alert_factor_est, alert_factor_SE, alert_factor_Pval_Estimate)
i3<- i3 %>% select(SNP, sleepiness_est, sleepiness_SE, sleepiness_Pval_Estimate)
i4<- i4 %>% select(SNP, diurnal_est, diurnal_SE, diurnal_Pval_Estimate)
i5<- i5 %>% select(SNP, napping_est, napping_SE, napping_Pval_Estimate)

tmp1<- merge(sigQ,f2, by='SNP'); nrow(tmp1)
tmp1<- merge(tmp1,i3, by='SNP'); nrow(tmp1)
tmp1<- merge(tmp1,i4, by='SNP'); nrow(tmp1)
sigQ<- merge(tmp1,i5, by='SNP'); nrow(sigQ)


sigQ$sigSleepy<- ifelse(sigQ$sleepiness_Pval_Estimate<5e-8,1,0)
sigQ$sigDiurn<- ifelse(sigQ$diurnal_Pval_Estimate<5e-8,1,0)
sigQ$sigNapping<- ifelse(sigQ$napping_Pval_Estimate<5e-8,1,0)

# write it in case it crashes
fwrite(sigQ, "Q_sig.txt", sep='\t')


# 21 significant SNPs from alertness GWAS that are not acting through the factor

sigAlert<- subset(alertQ, alertQ$Q_chisq_pval<5e-8)


```

circadian preference

```{r}
rm(list = ls(all.names = TRUE))

sigQ<- fread("Q_sig.txt", header=T, data.table=F)

f3<- fread("F3_Q.txt", header=T, data.table=F)
i6<- fread("indicator6.txt", header=T, data.table=F)
i7<- fread("indicator7.txt", header=T, data.table=F)
i8<- fread("indicator8.txt", header=T, data.table=F)
i9<- fread("indicator9.txt", header=T, data.table=F)

colnames(f3)[2:ncol(f3)]<- paste("chrono_factor", colnames(f3)[2:ncol(f3)], sep="_")
colnames(i6)[2:10]<- paste("chron", colnames(i6)[2:10], sep="_")
colnames(i7)[2:10]<- paste("mid", colnames(i7)[2:10], sep="_")
colnames(i8)[2:10]<- paste("m10", colnames(i8)[2:10], sep="_")
colnames(i9)[2:10]<- paste("l5", colnames(i9)[2:10], sep="_")


f3<- f3 %>% select(SNP, chrono_factor_est, chrono_factor_SE, chrono_factor_Pval_Estimate)
i6<- i6 %>% select(SNP, chron_est, chron_SE, chron_Pval_Estimate)
i7<- i7 %>% select(SNP, mid_est, mid_SE, mid_Pval_Estimate)
i8<- i8 %>% select(SNP, m10_est, m10_SE, m10_Pval_Estimate)
i9<- i9 %>% select(SNP, l5_est, l5_SE, l5_Pval_Estimate)

tmp1<- merge(sigQ,f3, by='SNP'); nrow(tmp1)
tmp1<- merge(tmp1,i6, by='SNP'); nrow(tmp1)
tmp1<- merge(tmp1,i7, by='SNP'); nrow(tmp1)
tmp1<- merge(tmp1,i8, by='SNP'); nrow(tmp1)
sigQ<- merge(tmp1,i9, by='SNP'); nrow(sigQ)

sigQ$sigChron<- ifelse(sigQ$chron_Pval_Estimate<5e-8,1,0)
sigQ$sigMid<- ifelse(sigQ$mid_Pval_Estimate<5e-8,1,0)
sigQ$sigM10<- ifelse(sigQ$m10_Pval_Estimate<5e-8,1,0)
sigQ$sigL5<- ifelse(sigQ$l5_Pval_Estimate<5e-8,1,0)

fwrite(sigQ, "Q_sig.txt", sep='\t')

```

we have 1322 significant Q TESTS. of those, which were significant SNPs for a factor?

```{r}

## effic
efficsig<- fread("/pl/active/friedman/studies/claire_ldsc_gsem/sleep_rr_upd/FUMA_out/effic/IndSigSNPs.txt", header=T, data.table=F)

sigQ$IndSig_Effic_F[sigQ$SNP %in% efficsig$rsID] <- 1  ### 1 sig: rs182588061

## alert
alertsig<- fread("/pl/active/friedman/studies/claire_ldsc_gsem/sleep_rr_upd/FUMA_out/alertness/IndSigSNPs.txt", header=T, data.table=F)

sigQ$IndSig_Alert_F[sigQ$SNP %in% alertsig$rsID] <- 1 ### 1 sig: rs72820274

## circ
circsig<- fread("/pl/active/friedman/studies/claire_ldsc_gsem/sleep_rr_upd/FUMA_out/circadian/IndSigSNPs.txt", header=T, data.table=F)

sigQ$IndSig_Circ_F[sigQ$SNP %in% circsig$rsID] <- 1   ### 2 sig: rs12654450 & rs182588061

fwrite(sigQ, "Q_sig.txt", sep='\t')

```

## Q SNP test
Take the 3 significant Q SNPs for factors, have them predict just that factor, then all indicators and compare that to the full model where SNPs predict all factors

Sig SNPs: 
rs12654450 = sig for circ
rs182588061 = sig for circ and effic
rs72820274 = sig for alert

```{r}
ml anaconda
conda activate /projects/lessem/software/anaconda/envs/R-4.2.1

R

library(devtools)
require(GenomicSEM)
library(dplyr)

load("/pl/active/friedman/studies/claire_ldsc_gsem/gsem/sleep_sumstats_06_2022.RData")
load("/pl/active/friedman/studies/claire_ldsc_gsem/gsem/sleepLDSC_GWAS_06_2022.RData")

circ_SNPs<- sleep_sumstats %>% filter(SNP=='rs12654450' | SNP=='rs182588061')

model<-"
F1=~efficiency+episodesR
F2=~sleepinessR+diurnalR+nappingR
F3=~chron+midR+m10R+l5R
F4=~self_sleepdur
F5=~sd_sleepdurR
F6=~insomR

sd_sleepdurR~~0*sd_sleepdurR
insomR~~0*insomR
self_sleepdur~~0*self_sleepdur

efficiency~~a*efficiency
a>.001


chron+midR+m10R+l5R~SNP
F1+F2+F4+F5+F6~SNP
"

circadianQ <- userGWAS(covstruc=sleepLDSC, SNPs=circ_SNPs, model=model, sub=c("chron~SNP", "midR~SNP", "m10R~SNP", "l5R~SNP"), parallel=F)


### save sumstats file
save(circadianQ, file="Circadian_Q_tests.RData")
```


THIS ONE WONT RUN!
this SNP runs on other models....
no other SNPs will run on this model

- per naomi: try setting the loading of episodes to the unst estimate from the model

```{r}
effic_SNPs<- sleep_sumstats %>% filter(SNP=='rs182588061')

model<-"
F1=~efficiency+0.15114535144*episodesR
F2=~sleepinessR+diurnalR+nappingR
F3=~chron+midR+m10R+l5R
F4=~self_sleepdur
F5=~sd_sleepdurR
F6=~insomR

sd_sleepdurR~~0*sd_sleepdurR
insomR~~0*insomR
self_sleepdur~~0*self_sleepdur

efficiency~~a*efficiency
a>.001


efficiency+episodesR~SNP
F2+F3+F4+F5+F6~SNP
"

efficiencyQ <- userGWAS(covstruc=sleepLDSC, SNPs=effic_SNPs, model=model, sub=c("efficiency~SNP","episodesR~SNP"), parallel=F)

save(efficiencyQ, file="../gsem/Efficiency_Q_tests.RData")

```

```{r}
alert_SNPs<- sleep_sumstats %>% filter(SNP=='rs72820274')

model<-"
F1=~efficiency+episodesR
F2=~sleepinessR+diurnalR+nappingR
F3=~chron+midR+m10R+l5R
F4=~self_sleepdur
F5=~sd_sleepdurR
F6=~insomR

sd_sleepdurR~~0*sd_sleepdurR
insomR~~0*insomR
self_sleepdur~~0*self_sleepdur

efficiency~~a*efficiency
a>.001


sleepinessR+diurnalR+nappingR~SNP
F1+F3+F4+F5+F6~SNP
"

alertnessQ <- userGWAS(covstruc=sleepLDSC, SNPs=alert_SNPs, model=model, sub=c("sleepinessR~SNP","diurnalR~SNP","nappingR~SNP"), parallel=F)

save(alertnessQ, file="Alertness_Q_tests.RData")
```



# compute differnece tests comparing model where SNP predicted ALL indicators to model where SNP predicted candidate factor *indicators* and all other factors

```{r}
ml anaconda
conda activate /projects/lessem/software/anaconda/envs/R-4.2.1

R


# re run the GWAS where each SNP predicts each indicator, just with the 3 SNPs


library(devtools)
require(GenomicSEM)
library(dplyr)


load("/pl/active/friedman/studies/claire_ldsc_gsem/gsem/sleep_sumstats_06_2022.RData")
load("/pl/active/friedman/studies/claire_ldsc_gsem/gsem/sleepLDSC_GWAS_06_2022.RData")

QSNPS<- sleep_sumstats %>% filter(SNP=="rs182588061" | SNP=='rs12654450' | SNP=='rs72820274')

model<-"
F1=~efficiency+episodesR
F2=~sleepinessR+diurnalR+nappingR
F3=~chron+midR+m10R+l5R
F4=~self_sleepdur
F5=~sd_sleepdurR
F6=~insomR

sd_sleepdurR~~0*sd_sleepdurR
insomR~~0*insomR
self_sleepdur~~0*self_sleepdur

efficiency~~a*efficiency
a>.001


efficiency+episodesR+sleepinessR+diurnalR+nappingR+chron+midR+m10R+l5R+self_sleepdur+sd_sleepdurR+insomR~SNP
"

CommonFactor <- userGWAS(covstruc=sleepLDSC, SNPs=QSNPS, model=model, sub=c("efficiency~SNP", "episodesR~SNP", "sleepinessR~SNP", "diurnalR~SNP", "nappingR~SNP", "chron~SNP", "midR~SNP", "m10R~SNP", "l5R~SNP"), parallel=F)

### just need model fit. should be same for each indicator

QSNPS<- CommonFactor[[1]]
QSNPS <- QSNPS %>% select(SNP, chisq, chisq_df, chisq_pval)


library(dplyr)
library(data.table)
sigQ<- fread("/pl/active/friedman/studies/claire_ldsc_gsem/", header=T, data.table=F)

colnames(sigQ)[2:ncol(sigQ)]<- paste0('indpath_', colnames(sigQ)[2:ncol(sigQ)])

## efficiency

## circadian 
load("Circadian_Q_tests.RData")

#chrono
circ<- (circadianQ)[[1]]
sigQ %>% right_join(circ) %>%
  mutate(chi_diff= chisq - indpath_chisq,
         df= chisq_df - indpath_chisq_df,
         Q_pval= pchisq(chi_diff,df,lower.tail=FALSE)) %>%
           select(SNP, chi_diff, df, Q_pval, est, SE, Pval_Estimate)
#mid
mid<- (circadianQ)[[2]]
sigQ %>% right_join(mid) %>%
  mutate(chi_diff= chisq - indpath_chisq,
         df= chisq_df - indpath_chisq_df,
         Q_pval= pchisq(chi_diff,df,lower.tail=FALSE)) %>%
           select(SNP, chi_diff, df, Q_pval, est, SE, Pval_Estimate)

#m10
m10<- (circadianQ)[[3]]
sigQ %>% right_join(m10) %>%
  mutate(chi_diff= chisq - indpath_chisq,
         df= chisq_df - indpath_chisq_df,
         Q_pval= pchisq(chi_diff,df,lower.tail=FALSE)) %>%
           select(SNP, chi_diff, df, Q_pval, est, SE, Pval_Estimate)

#l5
l5<- (circadianQ)[[4]]
sigQ %>% right_join(l5) %>%
  mutate(chi_diff= chisq - indpath_chisq,
         df= chisq_df - indpath_chisq_df,
         Q_pval= pchisq(chi_diff,df,lower.tail=FALSE)) %>%
           select(SNP, chi_diff, df, Q_pval, est, SE, Pval_Estimate)

## alertness 
load("Alertness_Q_tests.RData")

#sleepiness
sleepy<- (alertnessQ)[[1]]
sigQ %>% right_join(sleepy) %>%
  mutate(chi_diff= chisq - indpath_chisq,
         df= chisq_df - indpath_chisq_df,
         Q_pval= pchisq(chi_diff,df,lower.tail=FALSE)) %>%
           select(SNP, chi_diff, df, Q_pval,est, SE, Pval_Estimate)

#diurnal
dirun<- (alertnessQ)[[2]]
sigQ %>% right_join(dirun) %>%
  mutate(chi_diff= chisq - indpath_chisq,
         df= chisq_df - indpath_chisq_df,
         Q_pval= pchisq(chi_diff,df,lower.tail=FALSE)) %>%
           select(SNP, chi_diff, df, Q_pval,est, SE, Pval_Estimate)

#napping
napping<- (alertnessQ)[[3]]
sigQ %>% right_join(napping) %>%
  mutate(chi_diff= chisq - indpath_chisq,
         df= chisq_df - indpath_chisq_df,
         Q_pval= pchisq(chi_diff,df,lower.tail=FALSE)) %>%
           select(SNP, chi_diff, df, Q_pval,est, SE, Pval_Estimate)




## efficiency

load("/pl/active/friedman/studies/claire_ldsc_gsem/gsem/Efficiency_Q_tests.RData")

#effic
effic<- (efficiencyQ)[[1]]
sigQ %>% right_join(effic) %>%
  mutate(chi_diff= chisq - indpath_chisq,
         df= chisq_df - indpath_chisq_df,
         Q_pval= pchisq(chi_diff,df,lower.tail=FALSE)) %>%
           select(SNP, chi_diff, df, Q_pval,est, SE, Pval_Estimate)

#episodes
ep<- (efficiencyQ)[[2]]
sigQ %>% right_join(ep) %>%
  mutate(chi_diff= chisq - indpath_chisq,
         df= chisq_df - indpath_chisq_df,
         Q_pval= pchisq(chi_diff,df,lower.tail=FALSE)) %>%
           select(SNP, chi_diff, df, Q_pval,est, SE, Pval_Estimate)

```


Test chi squares

**Circadian:**

Chronotype:

          SNP chi_diff df       Q_pval         est          SE Pval_Estimate
1  rs12654450 249.6225  8 2.071339e-49 -0.01289400 0.001659660  7.907060e-15**
2 rs182588061 204.0239  8 9.065691e-40  0.01520531 0.006258893  1.512415e-02

Mid:

          SNP chi_diff df       Q_pval          est          SE Pval_Estimate
1  rs12654450 249.6225  8 2.071339e-49 -0.009815363 0.002982896  0.0009999241
2 rs182588061 204.0239  8 9.065691e-40  0.020263099 0.010800413  0.0606358906

m10:

          SNP chi_diff df       Q_pval         est          SE Pval_Estimate
1  rs12654450 249.6225  8 2.071339e-49 -0.00548685 0.002915713    0.05986031
2 rs182588061 204.0239  8 9.065691e-40  0.02096874 0.010614541    0.04821444


l5:

          SNP chi_diff df       Q_pval          est          SE Pval_Estimate
1  rs12654450 249.6225  8 2.071339e-49 -0.006046014 0.002933418  3.929521e-02
2 rs182588061 204.0239  8 9.065691e-40  0.086586135 0.009538527  1.110805e-19**


**Alertness:**

sleepiness:

        SNP chi_diff df       Q_pval         est           SE Pval_Estimate
1 rs72820274 372.8618  9 8.383831e-75 0.004081925 0.0005779611  1.633815e-12**

diurnal:

         SNP chi_diff df       Q_pval        est          SE Pval_Estimate
1 rs72820274 372.8618  9 8.383831e-75 0.00110336 0.002608328      0.672285

napping:

         SNP chi_diff df       Q_pval         est           SE Pval_Estimate
1 rs72820274 372.8618  9 8.383831e-75 0.002219132 0.0006882893   0.001263577

**Efficiency**

effic:

          SNP chi_diff df        Q_pval        est         SE Pval_Estimate
1 rs182588061 941.4413 11 7.609536e-195 0.07912458 0.00945686  5.915513e-17

episodes:

          SNP chi_diff df        Q_pval        est         SE Pval_Estimate
1 rs182588061 941.4413 11 7.609536e-195 0.03189478 0.01017048   0.001712603




## Q SNP comparisons
```{r}
library(dplyr)
library(data.table)
library(ggplot2)
library(ggnewscale)
q<- fread("/Users/claire/Desktop/sleepGWAS/Q/Q_sig.txt", header=T, data.table=F)
head(q)
nrow(q)

efficloading<- .357
eploading<- 1.0

sleepyloading<- .724
diurnalloading<- .366 
nappingloading<- .949

chronloading<- .902
midloading<- .941
m10loading<- .929
l5loading<- .955

q<- q %>% filter(IndSig_Effic_F==1 | IndSig_Alert_F==1 | IndSig_Circ_F==1)

### efficiency

ep<- q %>% mutate(implied_est= eff_factor_est*eploading,
                     diff=implied_est-episodes_est,
                  indicator='episodes') %>%
  select(SNP,episodes_est,episodes_SE,implied_est,diff,indicator) %>%
  rename(beta=episodes_est, SE=episodes_SE)

eff<- q %>% mutate(implied_est= eff_factor_est*efficloading,
                     diff=implied_est-efficiency_est,
                  indicator='efficiency') %>%
  select(SNP,efficiency_est,efficiency_SE,implied_est,diff,indicator) %>%
  rename(beta=efficiency_est, SE=efficiency_SE)


### alertness 
sleepy<- q %>% mutate(implied_est= alert_factor_est*sleepyloading,
                     diff=implied_est-sleepiness_est,
                     indicator="sleepiness") %>%
  select(SNP,sleepiness_est,sleepiness_SE,implied_est,diff,indicator)%>%
  rename(beta=sleepiness_est, SE=sleepiness_SE)

napping<- q %>% mutate(implied_est= alert_factor_est*nappingloading,
                     diff=implied_est-napping_est,
                     indicator="napping") %>%
  select(SNP,napping_est,napping_SE,implied_est,diff,indicator) %>%
  rename(beta=napping_est, SE=napping_SE)

diurnal<- q %>% mutate(implied_est= alert_factor_est*diurnalloading,
                     diff=implied_est-diurnal_est,
                     indicator="diurnal") %>%
  select(SNP,diurnal_est,diurnal_SE,implied_est,diff,indicator) %>%
  rename(beta=diurnal_est, SE=diurnal_SE)


### circadian preference

chron<- q %>% mutate(implied_est= chrono_factor_est*chronloading,
                     diff=implied_est-chron_est,
                     indicator="chronotype") %>%
  select(SNP,chron_est,chron_SE, implied_est,diff, indicator) %>%
  rename(beta=chron_est, SE=chron_SE)

mid<- q %>% mutate(implied_est= chrono_factor_est*midloading,
                     diff=implied_est-mid_est,
                     indicator="mid") %>%
  select(SNP,mid_est,mid_SE,implied_est,diff,indicator) %>%
  rename(beta=mid_est, SE=mid_SE)

m10<- q %>% mutate(implied_est= chrono_factor_est*m10loading,
                     diff=implied_est-m10_est,
                     indicator="m10") %>%
  select(SNP,m10_est,m10_SE,implied_est,diff,indicator) %>%
  rename(beta=m10_est, SE=m10_SE)

l5<- q %>% mutate(implied_est= chrono_factor_est*l5loading,
                     diff=implied_est-l5_est,
                     indicator="l5") %>%
  select(SNP,l5_est,l5_SE,implied_est,diff,indicator) %>%
  rename(beta=l5_est, SE=l5_SE)


```



Compute CIs for significant SNPs -- see if the SNP-->factor * indicator loading falls within CI
```{r}
dat<- rbind(eff, ep, sleepy, napping, diurnal, chron, m10, l5, mid)
dat

dat$`Observed Q SNP Effect`<- as.factor(dat$indicator)

plt<- dat %>% mutate(`Observed Q SNP Effect` = fct_relevel(`Observed Q SNP Effect`, 
            "chronotype", "l5", "m10", "mid",
            "sleepiness", "diurnal", "napping",
            "efficiency", "episodes"))

p<- ggplot(plt, aes(x=SNP, y=beta, fill=`Observed Q SNP Effect`)) + 
    geom_bar(position=position_dodge(), stat="identity",
             colour="black", # Use black outlines,
             size=.3) +      # Thinner lines
    geom_errorbar(aes(ymin=beta-SE, ymax=beta+SE),
                  size=.3,    # Thinner lines
                  width=.2,
                  position=position_dodge(.9)) +
  scale_colour_manual("4 cylinder", values = c("darkgrey")) +
    xlab("SNP") +
    ylab("Beta") +
    ggtitle("Observed vs Expected Q SNP Effects") +
    geom_point(aes(x=SNP, y=implied_est), size=4, color='darkgrey', position=position_dodge(.9),show.legend = F) +
  theme(plot.title = element_text(hjust = 0.5),
      panel.border = element_blank(),
      panel.grid.major = element_blank(),
      panel.grid.minor = element_blank(),
      axis.text.x = element_text(size=8))

ggsave(plot = p,filename = "/Users/claire/Desktop//Desktop/sleepGWAS/figs/expected_obs_QSNP.png", width = 9, height = 7, units = "in")


dat$Z<- abs(dat$beta/dat$SE)
dat$p<- pnorm(dat$Z, lower.tail = F)

dat %>% filter(p<5e-8)

```

Basically SNP rs____061 is Q for efficiency (overestimated) and L5 (overestimated)
SNP rs____274 is Q for sleepiness 
and 
SNP rs____450 is Q for chronotype


These plots show the Q SNPs are related to the indicators but are just not quite proportional. If we want to present, should probably re-order so the indicators are in the order of the factors. SNP rs____061 is greatly underestimated for L5, but over estimated for chronotype, in line with the conclusion that that SNP is a Q SNP for Circadian Preference, or not acting proportionally through the factor. For SNP rs______074, we don't see as much over/under estimation in the SNP effects. Same with rs____450, except that the SNP effect is under estimated for diurnal. Should look at the significance of the effect sizes for SNP--> indicator (observed SNP effects).

The chi square differences tests for both Q SNP tests for Circadian Preference were highly significant (chidiff(8)>204.03, ps<9.061531e-40). SNP rs___061 is driven by L5 (beta= 0.086586135, p=1.110808e-19) and partially by chronotype (beta= 0.01520531, p= 1.512411e-02). SNP rs____450 is driven mostly by chronoype (beta= -0.01289402, p= 7.906264e-15) and partially by sleep mid (beta= -0.009815383, p= 0.0009999004).

The chi square difference test for the Alertness Q SNP was highly signicant (chidiff(9)=372.86, p=8.383822e-75). SNP rs____274 is driven mostly by sleepiness (beta= 0.004081925, p= 1.633814e-12) and partially by napping (beta= 0.002219132, p= 0.001263577).

From katerina: removing SNPs in LD with our Q SNPs
(r2>.6)


```{r}
# QSNPS.tab simply lists all 1322 signifiant Q SNP tests

#Filter the files with LD estimates for Q SNPs: if column 3 of file 2 is in file 1, print:
for i in {1..22}
do
awk -F '\t' 'ARGIND==1{x[$0]=1;}ARGIND==2&&($3 in x){print $0;}' QSNPS.tab /pl/active/IBG/UKBiobank/computed_scores/LDscores/stats_ld_chr$i.ld1 > Qlist/Qsnp_stats_ld_$i.list
done

#Print the list of SNPs that are in LD at r2 > .6 with Q SNPs:
for i in {1..22}
do
awk '{if ($7>0.6) print $6;}' Qlist/Qsnp_stats_ld_$i.list > Qexclude/exclude_LDwQSNPs_ch$i.stats_ld.list
done

cat exclude_LDwQSNPs_ch* >> Qexclude.txt


# remove those SNPs from GWAS sum stats and run through FUMA
# /pl/active/friedman/studies/claire_ldsc_gsem/sleep_rr_upd/
  
ex<- fread("/pl/active/friedman/studies/claire_ldsc_gsem/sleep_rr_upd/Qexclude.txt", header=F, data.table=F)
ex<- ex$V1

circ<- fread("/pl/active/friedman/studies/claire_ldsc_gsem/sleep_rr_upd/circadian_pref_gsem.txt", header=T, data.table=F)
circ$inc<- ifelse(circ$SNP %in% ex, 0, 1)
circ<- subset(circ, circ$inc==1)
circ$inc<- NULL
fwrite(circ, "/pl/active/friedman/studies/claire_ldsc_gsem/sleep_rr_upd/circadian_pref_gsem.txt", sep='\t')


alert<- fread("/pl/active/friedman/studies/claire_ldsc_gsem/sleep_rr_upd/alertness_gsem.txt", header=T, data.table=F)
alert$inc<- ifelse(alert$SNP %in% ex, 0, 1)
alert<- subset(alert, alert$inc==1)
alert$inc<- NULL
fwrite(alert, "/pl/active/friedman/studies/claire_ldsc_gsem/sleep_rr_upd/alertness_gsem.txt", sep='\t')

eff<- fread("/pl/active/friedman/studies/claire_ldsc_gsem/sleep_rr_upd/efficiency_gsem.txt", header=T, data.table=F)
eff$inc<- ifelse(eff$SNP %in% ex, 0, 1)
eff<- subset(eff, eff$inc==1)
eff$inc<- NULL
fwrite(eff, "/pl/active/friedman/studies/claire_ldsc_gsem/sleep_rr_upd/efficiency_gsem.txt", sep='\t')


```

### get data ready for massive

```{r}
library(data.table)
library(dplyr)

# alert
alert<- fread("/Users/clairemorrison/Desktop/sleepGWAS/sleep_rr_upd/alertness_gsem.txt.gz", header=T, data.table=F)
head(alert)


alert<- alert %>% select(CHR, BP, SNP, A1, A2, MAF, est, SE, Pval_Estimate) %>%
  rename(FREQ = MAF,
         BETA = est,
         P = Pval_Estimate)

head(alert)

fwrite(alert, "/Users/clairemorrison/Desktop/sleepGWAS/MASSIVE/alertness.txt", sep='\t')

# circ
circ<- fread("/Users/clairemorrison/Desktop/sleepGWAS/sleep_rr_upd/circadian_pref_gsem.txt.gz", header=T, data.table=F)
head(circ)


circ<- circ %>% select(CHR, BP, SNP, A1, A2, MAF, est, SE, Pval_Estimate) %>%
  rename(FREQ = MAF,
         BETA = est,
         P = Pval_Estimate)

head(circ)

fwrite(circ, "/Users/clairemorrison/Desktop/sleepGWAS/MASSIVE/circadian_pref.txt", sep='\t')


# effic
effic<- fread("/Users/clairemorrison/Desktop/sleepGWAS/sleep_rr_upd/efficiency_gsem.txt.gz", header=T, data.table=F)
head(effic)


effic<- effic %>% select(CHR, BP, SNP, A1, A2, MAF, est, SE, Pval_Estimate) %>%
  rename(FREQ = MAF,
         BETA = est,
         P = Pval_Estimate)

head(effic)

fwrite(effic, "/Users/clairemorrison/Desktop/sleepGWAS/MASSIVE/efficiency.txt", sep='\t')


# insom
insom<- fread("/Users/clairemorrison/Desktop/sleepGWAS/sleep_rr_upd/insomnia_withNeffR.txt.gz", header=T, data.table=F)
head(insom)


insom<- insom %>% select(CHR, BP, SNP, A1, A2, MAF, beta, se, p) %>%
  rename(FREQ = MAF,
         BETA = beta,
         P = p,
         SE=se)

head(insom)

fwrite(insom, "/Users/clairemorrison/Desktop/sleepGWAS/CTG-VL/MASSIVE//insomnia.txt", sep='\t')

# dur
dur<- fread("/Users/clairemorrison/Desktop/sleepGWAS/sleep_rr_upd/selfsleepdur.txt.gz", header=T, data.table=F)
head(dur)


dur<- dur %>% select(CHR, BP, SNP, A1, A2, MAF, beta, se, p) %>%
  rename(FREQ = MAF,
         BETA = beta,
         P = p,
         SE=se)

head(dur)

fwrite(dur, "/Users/clairemorrison/Desktop/sleepGWAS//CTG-VL/MASSIVE/duration.txt", sep='\t')


# var
var<- fread("/Users/clairemorrison/Desktop/sleepGWAS/sleep_rr_upd/sdsleepdurR.txt.gz", header=T, data.table=F)
head(var)


var<- var %>% select(CHR, BP, SNP, A1, A2, MAF, beta, SE, p) %>%
  rename(FREQ = MAF,
         BETA = beta,
         P = p)


head(var)

fwrite(var, "/Users/clairemorrison/Desktop/sleepGWAS/CTG-VL//MASSIVE/variability.txt", sep='\t')


```

# post FUMA/MAGMA comparisons

### Step 8: compare sig SNPs in our GWAS to sig SNPs in component GWAS
/pl/active/friedman/studies/claire_ldsc_gsem/SLEEP_6_FAC_GWAS/gsem_gwas_out has GWAS out files
/pl/active/friedman/studies/claire_ldsc_gsem/ldsc/sleepsumstats/cleaned has component GWAS out files
/pl/active/friedman/studies/claire_ldsc_gsem/SLEEP_6_FAC_GWAS has file with signifciant component GWAS SNPs

/pl/active/friedman/studies/claire_ldsc_gsem/SLEEP_6_FAC_GWAS/FUMA_out has FUMA output folders



# Upset plot of common SNPs from component GWAS'

```{r}
library(dplyr)
library(tidyr)
library(tidyverse)
library(ComplexUpset)


sleep_comp<- fread("/Users/claire/Desktop/Desktop/sleepGWAS/sleep_comp_SNPs.csv", header=T, data.table=F)
length(unique(sleep_comp$SNP))


traits <- c("Chronotype", "Mid", "M10", "L5", "Alertness", "Diurnal", "Efficiency", "Episodes", "SleepDurSelf", "SleepDurActi", "Insomnia")
names(traits) <- traits

dat <- readxl::read_xlsx("/Users/claire/Desktop//Desktop/sleepGWAS/SNPoverlap.xlsx") 
dat %>% print(n = nrow(dat))

subsets <- dat$combination

symptom_mat <- map_dfc(subsets, str_detect, traits) %>%
    data.frame() %>%
    t() %>% # transpose the result, ugh
    as_tibble()
colnames(symptom_mat)  <- traits
symptom_mat$count <- dat$count
symptom_mat %>% print(n = nrow(symptom_mat))


indvs <- symptom_mat %>%
    uncount(count) 
indvs


upset(data = indvs, intersect = traits, 
      name="SNP hits by Frequency.", 
      min_size = 0,
      width_ratio = 0.125) +
    labs(title = "Overlap of Genome-Wide Signal Between Sleep Traits")

```





look at z score > 5.326762 aka p val < 5e-8
```{r chronotype}
library(data.table)
library(dplyr)

### circadiam
cir<- fread("/Users/claire/Desktop//Desktop/sleepGWAS/sleep_rr_upd//circadian_pref_gsem.txt.gz", header=T, data.table=F)
nrow(cir) # 7409883

#cir<- subset(cir, cir$Z_Estimate>5.326762)

# only genomic risk loci
cir_loci<- fread("/Users/claire/Desktop//Desktop/sleepGWAS//FUMA_out/circadian/GenomicRiskLoci.txt", header=T, data.table=F)
nrow(cir_loci)
cir_loci<- merge(cir, cir_loci, by.x='SNP', by.y='rsID', all.y=T)  ### 98 significant genomic risk loci 


# with chronotype
comp_chron<- fread("/Users/claire/Desktop//Desktop/sleepGWAS/component/chrono_comp/GenomicRiskLoci.txt", header=T, data.table=F)
nrow(comp_chron) #191 sig loci in chronotype indicator GWAS
comp_chron<- comp_chron %>% select(rsID) %>%
  rename(SNP=rsID)

cir_loci %>% inner_join(comp_chron) %>%
  select(SNP) ### 30 SNPs --> 191-30= 161 SNPs not found in our GWAS


# wtih m10
comp_m10<- fread("/Users/claire/Desktop//Desktop/sleepGWAS/component/m10/GenomicRiskLoci.txt", header=T, data.table=F)
comp_m10<- comp_m10 %>% select(rsID) %>%
  rename(SNP=rsID)
nrow(comp_m10) # 1 sig SNP

cir_loci %>% inner_join(comp_m10) %>%
  select(SNP) ### 0 SNPs --> so 1 sig SNP not found in our GWAS

# with l5
comp_l5<- fread("/Users/claire/Desktop//Desktop/sleepGWAS/component/l5/GenomicRiskLoci.txt", header=T, data.table=F)
comp_l5<- comp_l5 %>% select(rsID) %>%
  rename(SNP=rsID)

nrow(comp_l5) # 6 sig SNPs --> 4 not found by our factor

cir_loci %>% inner_join(comp_l5) %>%
  select(SNP) ### 2 SNPs


# with mid
comp_mid<- fread("/Users/claire/Desktop//Desktop/sleepGWAS/component/mid/GenomicRiskLoci.txt", header=T, data.table=F)
comp_mid<- comp_mid %>% select(rsID) %>%
  rename(SNP=rsID)
nrow(comp_mid) # 1 sig SNP

cir_loci %>% inner_join(comp_mid) %>%
  select(SNP) ### 0 SNPs --> so 1 not found by our factor


# are the L5 SNPs also found in Chronotype?
intersect(cir_loci$SNP, comp_l5$SNP) ### yes both


############## 

#find novel SNPs

d<- merge(comp_chron, comp_m10, all=T, by='SNP')
d2<- merge(d, comp_l5, all=T, by='SNP')
d3<- merge(d2, comp_mid, all=T, by='SNP')

chron_new<- as.data.frame(setdiff(cir_loci$SNP, d3$SNP))  ### 67 new SNPs
colnames(chron_new)[1]<- 'SNP'
nrow(chron_new) ### 67 new

fwrite(chron_new, "/Users/claire/Desktop//Desktop/sleepGWAS/circ_novel_SNPs.txt", sep='\t')


# find SNPs we did not get as hits

not_in_cir_fact<- data.frame()
cir_indicators<- c("comp_chron", "comp_m10", "comp_l5", "comp_mid")
for (i in cir_indicators) {
  dat<- eval(parse(text=i)) %>% anti_join(cir_loci) %>%
  mutate(indicator=i)
  not_in_cir_fact<- rbind(not_in_cir_fact, dat)
}

table(not_in_cir_fact$indicator) # checks out with calcs

fwrite(not_in_cir_fact, "/Users/claire/Desktop//Desktop/sleepGWAS/circ_SNPs_not_in_factor.txt", sep='\t')


```

```{r effic}
#rm(list = ls(all.names = TRUE))


### efficiency
eff<- fread("/Users/claire/Desktop/Desktop/sleepGWAS/sleep_rr_upd//efficiency_gsem.txt.gz", header=T, data.table=F)
nrow(eff) # 7409883


# only genomic risk loci
eff_loci<- fread("/Users/claire/Desktop/Desktop/sleepGWAS/FUMA_out/effic/GenomicRiskLoci.txt", header=T, data.table=F) ### 8 genomic risk loci
eff_loci<- merge(eff, eff_loci, by.x='SNP', by.y='rsID', all.y=T) # 8 

# with efficiency
comp_eff<- fread("/Users/claire/Desktop/Desktop/sleepGWAS/component/effic_comp/GenomicRiskLoci.txt", header=T, data.table=F)
comp_eff<- comp_eff %>% select(rsID) %>%
  rename(SNP=rsID)

eff_loci %>% inner_join(comp_eff) %>%
  select(SNP) ### 1 SNPs

# with episodes
comp_ep<- fread("/Users/claire/Desktop/Desktop/sleepGWAS/component/episodes/GenomicRiskLoci.txt", header=T, data.table=F)
comp_ep<- comp_ep %>% select(rsID) %>%
  rename(SNP=rsID)

eff_loci %>% inner_join(comp_ep) %>%
  select(SNP) ### 3 SNPs


# is the efficiency SNP also found in epsidoes
intersect(comp_eff$SNP, comp_ep$SNP) ### no

# how many SNPs for all the component GWAS'?

e<- merge(comp_eff, comp_ep, by='SNP', all=T)

effic_new<- as.data.frame(setdiff(eff_loci$SNP, e$SNP))  ### 4 new SNPs
nrow(effic_new)
colnames(effic_new)[1]<- 'SNP'
fwrite(effic_new, "/Users/claire/Desktop/Desktop/sleepGWAS/effic_novel_SNPs.txt", sep='\t')


# find SNPs we did not get as hits in our factor

not_in_eff_fact<- data.frame()
eff_indicators<- c("comp_ep", "comp_eff")
for (i in eff_indicators) {
  dat<- eval(parse(text=i)) %>% anti_join(eff_loci) %>%
  mutate(indicator=i) %>%
    select()
  not_in_eff_fact<- rbind(not_in_eff_fact, dat)
}

table(not_in_eff_fact$indicator)

fwrite(not_in_eff_fact, "/Users/claire/Desktop//Desktop/sleepGWAS/eff_SNPs_not_in_factor.txt", sep='\t')

```

```{r alertness}
rm(list = ls(all.names = TRUE))


### alertness
alert<- fread("/Users/claire/Desktop//Desktop/sleepGWAS/sleep_rr_upd//alertness_gsem.txt.gz", header=T, data.table=F)
nrow(alert) # 7409883

# only genomic risk loci
alert_loci<- fread("/Users/claire/Desktop//Desktop/sleepGWAS//FUMA_out/alertness/GenomicRiskLoci.txt", header=T, data.table=F)
alert_loci<- merge(alert, alert_loci, by.x='SNP', by.y='rsID', all.y=T) ### 85

# with alertness
comp_alert<- fread("/Users/claire/Desktop//Desktop/sleepGWAS/component/sleepiness/GenomicRiskLoci.txt", header=T, data.table=F)
comp_alert<- comp_alert %>% select(rsID) %>%
  rename(SNP=rsID)

alert_loci %>% inner_join(comp_alert) %>%
  select(SNP) ### 8 SNPs

# with napping
comp_nap<- fread("/Users/claire/Desktop//Desktop/sleepGWAS/component/napping/GenomicRiskLoci.txt", header=T, data.table=F)
comp_nap<- comp_nap %>% select(rsID) %>%
  rename(SNP=rsID)

alert_loci %>% inner_join(comp_nap) %>%
  select(SNP) ### 31 SNPs

# with diurnal
comp_diurn<- fread("/Users/claire/Desktop//Desktop/sleepGWAS/component/diurnal/GenomicRiskLoci.txt", header=T, data.table=F)
comp_diurn<- comp_diurn %>% select(rsID) %>%
  rename(SNP=rsID)

alert_loci %>% inner_join(comp_diurn) %>%
  select(SNP) ### 0 SNPs

# are the napping SNP also found in alertness
intersect(comp_nap$SNP, comp_alert$SNP) ### just 1 rs12140153

# how many SNPs for all the component GWAS'?

a<- merge(comp_alert, comp_nap, by='SNP', all=T)
a2<- merge(a, comp_diurn, by='SNP', all=T)
nrow(a2) ### 164

alert_new<- as.data.frame(setdiff(alert_loci$SNP, a2$SNP))  ### 47 new SNPs
nrow(alert_new)
colnames(alert_new)[1]<- 'SNP'

fwrite(alert_new, "/Users/claire/Desktop//Desktop/sleepGWAS/alert_novel_SNPs.txt", sep='\t')

# find SNPs we did not get as hits in our factor

not_in_alert_fact<- data.frame()
alert_indicators<- c("comp_nap", "comp_alert", "comp_diurn")
for (i in alert_indicators) {
  dat<- eval(parse(text=i)) %>% anti_join(alert_loci) %>%
  mutate(indicator=i)
  not_in_alert_fact<- rbind(not_in_alert_fact, dat)
}

table(not_in_alert_fact$indicator) # 

fwrite(not_in_alert_fact, "/Users/claire/Desktop//Desktop/sleepGWAS/alert_SNPs_not_in_factor.txt", sep='\t')

```



### number of SNPs in common across factors

```{r}
library(data.table)
library(dplyr)
alert<- fread("/Users/clairemorrison/Desktop/sleepGWAS/FUMA_out//alertness/GenomicRiskLoci.txt", header=T, data.table=T)
alert<- alert %>% select(rsID) %>%
  mutate(alert=1)

circ<- fread("/Users/clairemorrison/Desktop/sleepGWAS/FUMA_out/circadian/GenomicRiskLoci.txt", header=T, data.table=T)
circ<- circ %>% select(rsID) %>%
  mutate(circ=1)

eff<- fread("/Users/clairemorrison/Desktop/sleepGWAS/FUMA_out/effic/GenomicRiskLoci.txt", header=T, data.table=T)
eff<- eff %>% select(rsID) %>%
  mutate(eff=1)

insom<- fread("/Users/clairemorrison/Desktop/sleepGWAS/FUMA_out/insom/GenomicRiskLoci.txt", header=T, data.table=T)
insom<- insom %>% select(rsID) %>%
  mutate(insom=1)

dur<- fread("/Users/clairemorrison/Desktop/sleepGWAS/FUMA_out/duration/GenomicRiskLoci.txt", header=T, data.table=T)
dur<- dur %>% select(rsID) %>%
  mutate(dur=1)


d<- merge(alert,circ,by='rsID',all=T)
d<- merge(d,eff,by='rsID',all=T)
d<- merge(d,insom,by='rsID',all=T)
d<- merge(d,dur,by='rsID',all=T)


d<- d %>%
   replace(is.na(.), 0) %>%
   mutate(sum = rowSums(across(where(is.numeric))))

table(d$sum)
nrow(d)

# 311 SNPs associated with any factor -- 3 associated with 2
subset(d, d$sum>1) # SNPs associated with > 1 factor
#rs12140153 = alert and circ
#rs7556815 = insom and dur
#rs75606464 = insom and dur
```


### number of genes in common across factors

```{r}
rm(list = ls(all.names = TRUE))
library(data.table)
library(dplyr)

alert<- fread("/Users/claire//Desktop/sleepGWAS/FUMA_out/alertness/genes.txt", header=T, data.table=T)
alert<- alert %>% select(symbol) %>%
  mutate(alert=1)

circ<- fread("/Users/claire//Desktop/sleepGWAS/FUMA_out/circadian/genes.txt", header=T, data.table=T)
circ<- circ %>% select(symbol) %>%
  mutate(circ=1)

eff<- fread("/Users/claire//Desktop/sleepGWAS/FUMA_out/effic/genes.txt", header=T, data.table=T)
eff<- eff %>% select(symbol) %>%
  mutate(eff=1)

insom<- fread("/Users/claire//Desktop/sleepGWAS/FUMA_out/insom/genes.txt", header=T, data.table=T)
insom<- insom %>% select(symbol) %>%
  mutate(insom=1)

dur<- fread("/Users/claire//Desktop/sleepGWAS/FUMA_out/duration/genes.txt", header=T, data.table=T)
dur<- dur %>% select(symbol) %>%
  mutate(dur=1)


d<- merge(alert,circ,by='symbol',all=T)
d<- merge(d,eff,by='symbol',all=T)
d<- merge(d,insom,by='symbol',all=T)
d<- merge(d,dur,by='symbol',all=T)


d<- d %>%
   replace(is.na(.), 0) %>%
   mutate(sum = rowSums(across(where(is.numeric))))

table(d$sum) # 8 associated with 4 factors, 147 assoc with 3, 697 assoc with 2, 2539 assoc with 11
nrow(d) # 75% of genes assoc with any factor were only assoc with 1 factor

subset(d, d$sum>3)
```

### network plot
```{r}
alert<- alert %>% rename(from=alert, 
                         to=symbol) %>%
  mutate(from="Alertness")

circ<- circ %>% rename(from=circ, 
                         to=symbol) %>%
  mutate(from="Circadian Preference")

eff<- eff %>% rename(from=eff, 
                         to=symbol) %>%
  mutate(from="Efficiency")

insom<- insom %>% rename(from=insom, 
                         to=symbol) %>%
  mutate(from="Non-Insomnia")

dur<- dur %>% rename(from=dur, 
                         to=symbol) %>%
  mutate(from="Duration")


network<- rbind(alert, circ, eff, insom, dur)

head(network)

head(hierarchy)

vertices <- data.frame(name = unique(c(as.character(network$from), as.character(network$to))) ) 

head(vertices)

from <- match( network$from, vertices$name)
to <- match( network$to, vertices$name)


ggraph(mygraph, layout = 'dendrogram', circular = TRUE) + 
  geom_conn_bundle(data = get_con(from = from, to = to), alpha=0.2, colour="skyblue", tension = 0) + 
  geom_node_point(aes(filter = leaf, x = x*1.05, y=y*1.05)) +
  theme_void()
```


## do again but for upset plot

```{r}

library(dplyr)
library(tidyr)
library(tidyverse)
library(ComplexUpset)
library(data.table)
library(stringr)

alert<- fread("/Users/claire/Desktop//sleepGWAS/FUMA_out/alertness/genes.txt", header=T, data.table=T)
alert<- alert %>% select(symbol) %>%
  mutate(alert="Alertness")

circ<- fread("/Users/claire/Desktop//sleepGWAS/FUMA_out/circadian/genes.txt", header=T, data.table=T)
circ<- circ %>% select(symbol) %>%
  mutate(circ="Circadian Preference")

eff<- fread("/Users/claire/Desktop//sleepGWAS/FUMA_out/effic/genes.txt", header=T, data.table=T)
eff<- eff %>% select(symbol) %>%
  mutate(eff="Efficiency")

insom<- fread("/Users/claire/Desktop//sleepGWAS/FUMA_out/insom/genes.txt", header=T, data.table=T)
insom<- insom %>% select(symbol) %>%
  mutate(insom="Insomnia")

dur<- fread("/Users/claire/Desktop//sleepGWAS/FUMA_out/duration/genes.txt", header=T, data.table=T)
dur<- dur %>% select(symbol) %>%
  mutate(dur="Duration")

d<- merge(alert,circ,by='symbol',all=T)
d<- merge(d,eff,by='symbol',all=T)
d<- merge(d,insom,by='symbol',all=T)
d<- merge(d,dur,by='symbol',all=T)



d<- d %>% mutate(comb = paste(alert,circ,eff,insom,dur, sep="&"))

head(d)

### overlapping factors are:
# alert and circ
# alert and circ and insom
# alert and dur
# alert and insom
# circ and dur
# circ and effic and insom and dur
# circ and insom


d$comb2<- str_replace_all(d$comb, "NA", "")
d$comb3<- str_replace_all(d$comb2, c("&&&&"="&", "&&&"="&", "&&&"="&"))
d$comb4<- str_replace_all(d$comb3, "&&", "&")


lut<- c("&Circadian Preference&" = "Circadian Preference",
        "&Circadian Preference&Duration" = "Circadian Preference&Duration",
        "&Circadian Preference&Efficiency&" = "Circadian Preference&Efficiency",
        "&Circadian Preference&Efficiency&Insomnia&" = "Circadian Preference&Efficiency&Insomnia",
        "&Circadian Preference&Insomnia&" = "Circadian Preference&Insomnia",
        "&Circadian Preference&Insomnia&Duration" = "Circadian Preference&Insomnia&Duration",
        "&Duration" = "Duration",
        "&Efficiency&" = "Efficiency",
       # "&Efficiency&Insomnia&" = "Efficiency&Insomnia",
        "&Insomnia&" = "Insomnia",
        "&Insomnia&Duration" = "Insomnia&Duration",
        "Alertness&" = "Alertness",
        "Alertness&Circadian Preference&" = "Alertness&Circadian Preference",
        "Alertness&Circadian Preference&Duration" = "Alertness&Circadian Preference&Duration",
        "Alertness&Circadian Preference&Insomnia&" = "Alertness&Circadian Preference&Insomnia",
        "Alertness&Circadian Preference&Insomnia&Duration" = "Alertness&Circadian Preference&Insomnia&Duration",
        "Alertness&Duration" = "Alertness&Duration",
        "Alertness&Efficiency&" = "Alertness&Efficiency",
        "Alertness&Insomnia&" = "Alertness&Insomnia",
        "Alertness&Insomnia&Duration" = "Alertness&Insomnia&Duration")

        
head(d)

d$combination <- lut[d$comb4]

d<- d %>% select(symbol, combination) %>%
  group_by(combination) %>%
  tally()

traits <- c("Alertness", "Circadian Preference", "Duration", "Insomnia", "Efficiency")
names(traits) <- traits

d %>% print(n = nrow(d))

subsets <- d$combination

symptom_mat <- map_dfc(subsets, str_detect, traits) %>%
    data.frame() %>%
    t() %>% # transpose the result, ugh
    as_tibble()
colnames(symptom_mat)  <- traits
symptom_mat$n <- d$n
symptom_mat %>% print(n = nrow(symptom_mat))


indvs <- symptom_mat %>%
    uncount(n) 
indvs


p<- upset(data = indvs, intersect = traits, 
      name="Genes by Frequency.", 
      min_size = 0,
      width_ratio = 0.125) +
    labs(title = "Overlap of Significant Genes Between Sleep Health Factors")
  
ggsave(plot = p,filename = "/Users/claire/Desktop//Desktop/sleepGWAS/figs/gene_overlap.png", width = 9, height = 7, units = "in")

```

trying to plot as sankey

```{r}
library(networkD3)


alert<- fread("/Users/claire/Desktop//sleepGWAS/FUMA_out/alertness/genes.txt", header=T, data.table=T)
alert<- alert %>% select(symbol) %>%
  mutate(sleep="Alertness",
         source=0)

circ<- fread("/Users/claire/Desktop//sleepGWAS/FUMA_out/circadian/genes.txt", header=T, data.table=T)
circ<- circ %>% select(symbol) %>%
  mutate(sleep="Circadian Preference",
         source=1)

eff<- fread("/Users/claire/Desktop//sleepGWAS/FUMA_out/effic/genes.txt", header=T, data.table=T)
eff<- eff %>% select(symbol) %>%
  mutate(sleep="Efficiency",
         source=2)

insom<- fread("/Users/claire/Desktop//sleepGWAS/FUMA_out/insom/genes.txt", header=T, data.table=T)
insom<- insom %>% select(symbol) %>%
  mutate(sleep="Insomnia",
         source=3)

dur<- fread("/Users/claire/Desktop//sleepGWAS/FUMA_out/duration/genes.txt", header=T, data.table=T)
dur<- dur %>% select(symbol) %>%
  mutate(sleep="Duration",
         source=4)


genes<- rbind(alert, circ, eff, insom, dur)

sleep<- unique(genes$sleep)
genes_list<- data.frame(unique(genes$symbol))
colnames(genes_list)[1]<- 'genes'
genes_list$target<- seq(1,nrow(genes_list))

genes_list$target<- genes_list$target+4 ### no overlap between sleep keys


genes$target <- factor(genes$symbol, levels=genes_list$genes, labels=genes_list$target)
head(genes)

nodes_sleep<- genes %>% group_by(sleep) %>%
  slice(1) %>%
  select(source)

nodes_genes<- genes %>% group_by(symbol) %>%
  slice(1) %>%
  select(target)

names_sleep<- data.frame(names=nodes_sleep$sleep)
names_genes<- data.frame(names=nodes_genes$symbol)

nodes<- rbind(names_sleep, names_genes)


links<- genes

links$symbol<- NULL
links$sleep<- NULL

links$Value<- 1

#### trying another way

genes<- fread("/Users/claire/Desktop/sankey_genes.csv", header=T, data.table=F)
head(genes)

genes$from[genes$sleep=="circadian"]<- 0
genes$from[genes$sleep=="duration"]<- 1
genes$from[genes$sleep=="alertness"]<- 2
genes$from[genes$sleep=="insomnia"]<- 3
genes$from[genes$sleep=="efficiency"]<- 4

genes$to[genes$sleep2=="circadian"]<- 5
genes$to[genes$sleep2=="duration"]<- 6
genes$to[genes$sleep2=="alertness"]<- 7
genes$to[genes$sleep2=="insomnia"]<- 8
genes$to[genes$sleep2=="efficiency"]<- 9


nodes<- data.frame(names=rep(c("circadian", "duration", "alertness", "insomnia", "efficiency"),2))

links<- genes
links$sleep<- NULL
links$sleep2<- NULL


sankeyNetwork(Links = links, Nodes = nodes, Source = "from",
              Target = "to", Value = "value", NodeID = "names",
              units = "TWh", fontSize = 12, nodeWidth = 30)


### with three+ levels

genes<- fread("/Users/claire/Desktop/sankey_genes.csv", header=T, data.table=F)
head(genes)


nodes<- data.frame(names=genes$names)
nodes<- nodes[1:13,]
nodes<- data.frame(names=nodes)

links<- genes
links$sleep<- NULL
links$sleep2<- NULL


p<- sankeyNetwork(Links = links, Nodes = nodes, Source = "from",
              Target = "to", Value = "value", NodeID = "names",
              units = "genes", fontSize = 12, nodeWidth = 30)

library(htmlwidgets)
saveWidget(p, file="/Users/claire/Desktop/sankey.html")

### fewer levels i think -- deosnt like this one

genes<- fread("/Users/claire/Desktop/sankey_genes2.csv", header=T, data.table=F)
head(genes)


nodes<- data.frame(names=genes$names)
nodes<- nodes[1:10,]
nodes<- data.frame(names=nodes)

links<- genes
links$sleep<- NULL
links$sleep2<- NULL


sankeyNetwork(Links = links, Nodes = nodes, Source = "from",
              Target = "to", Value = "value", NodeID = "names",
              units = "genes", fontSize = 12, nodeWidth = 30)


```


```{r}
library(ggraph)
library(igraph)
library(tidyverse)
library(RColorBrewer)
 
# create a data frame giving the hierarchical structure of your individuals

d<- merge(alert,circ,by='symbol',all=T)
d<- merge(d,eff,by='symbol',all=T)
d<- merge(d,insom,by='symbol',all=T)
d<- merge(d,dur,by='symbol',all=T)

alert<- alert %>% rename(from=alert,
                     to=symbol)
circ<- circ %>% rename(from=circ,
                     to=symbol)
eff<- eff %>% rename(from=eff,
                     to=symbol)
insom<- insom %>% rename(from=insom,
                     to=symbol)
dur<- dur %>% rename(from=dur,
                     to=symbol)

t2<- rbind(alert,circ,eff,insom,dur)

t2<- data.frame(t2)

```

*1/6 update*

we are plotting the GO bio, molec and celluar terms
feedback from naomi: don't facet wrap. organize more by process?
aka add annotation lines maybe that say 'bio', 'molec' and 'cellular'

*1/20 update*

try to order factors in legend with bars in plot

### standardize colors per factor across all plots
```{r}
circadian_col<- "brown2"
alertness_col<- "#00BFC4"
efficiency_col<- "#619CFF"
duration_col<- "#FF61CC"
non_insom_col<- "goldenrod"
regularity_col<- "seagreen"
  
fact_colors<- c("Circadian Preference"=circadian_col, "Alertness"=alertness_col, "Efficiency"=efficiency_col, "Duration"=duration_col, "Non-Insomnia"=non_insom_col, "Regularity"=regularity_col)
```

## plot GO terms

reading in magma.gsa.out (not totally sure why) but it's a text file that has three lines of headers. I deleted the info lines

total genes tested per phenotype= 18761
bonf p=5.330206e-07?
magma is using the corrected p of 1e-3 though.... not sure why?

from Magma:
 Plots and tables only display gene sets with adjusted P-value < 0.05. When adjusted P-value threshold is set to > 0.05, all results passed threshold are included in the GS.txt field downloadable from "Summary" tab.
 If there is no significant gene sets (adjusted P-value < 0.05) in user provided custom gene sets, they are not displayed in this page, but all results passed threshold are included in the GS.txt field downloadable from "Summary" tab.

**REPLOT AT SOME POINT**

```{r}
library(stringr)
library(data.table)
library(dplyr)
library(ggplot2)


          
GObio<- c("GO_SKELETAL_SYSTEM_DEVELOPMENT",
          "GO_STRESS_RESPONSE_TO_METAL_ION",
          "GO_CELLULAR_RESPONSE_TO_ZINC_ION",
          "GO_CELLULAR_RESPONSE_TO_COPPER_ION",
          "GO_PYRIMIDINE_RIBONUCLEOSIDE_CATABOLIC_PROCESS",
          "GO_TELENCEPHALON_DEVELOPMENT",
          "GO_PYRIMIDINE_NUCLEOSIDE_CATABOLIC_PROCESS",
          "GO_ZINC_ION_HOMEOSTASIS",
          "GO_TRNA_AMINOACYLATION_FOR_MITOCHONDRIAL_PROTEIN_TRANSLATION",
          "GO_NEGATIVE_REGULATION_OF_GROWTH",
          "GO_CYTIDINE_TO_URIDINE_EDITING",
          "GO_DNA_REPLICATION_DEPENDENT_NUCLEOSOME_ORGANIZATION",
          "GO_CHROMATIN_ASSEMBLY",
          "GO_NUCLEOSOME_ORGANIZATION",
          "GO_CHROMATIN_ASSEMBLY_OR_DISASSEMBLY",
          "GO_DNA_PACKAGING",
          "GO_CHROMATIN_SILENCING_AT_RDNA",
          "GO_PROTEIN_HETEROTETRAMERIZATION",
          "GO_PROTEIN_DNA_COMPLEX_SUBUNIT_ORGANIZATION",
          "GO_DNA_CONFORMATION_CHANGE",
          "GO_NEGATIVE_REGULATION_OF_GENE_EXPRESSION_EPIGENETIC",
          "GO_REGULATION_OF_MEGAKARYOCYTE_DIFFERENTIATION",
          "GO_CHROMATIN_ORGANIZATION_INVOLVED_IN_REGULATION_OF_TRANSCRIPTION",
          "GO_REGULATION_OF_GENE_SILENCING",
          "GO_NEGATIVE_REGULATION_OF_MEGAKARYOCYTE_DIFFERENTIATION",
          "GO_MEGAKARYOCYTE_DIFFERENTIATION",
          "GO_CHROMATIN_ORGANIZATION",
          "GO_REGULATION_OF_POSTTRANSCRIPTIONAL_GENE_SILENCING",
          "GO_PROTEIN_TETRAMERIZATION",
          "GO_PROTEIN_HETEROOLIGOMERIZATION",
          "GO_CHROMOSOME_ORGANIZATION",
          "GO_CELLULAR_PROTEIN_CONTAINING_COMPLEX_ASSEMBLY",
          "GO_INTERLEUKIN_7_MEDIATED_SIGNALING_PATHWAY",
          "GO_TELOMERE_ORGANIZATION",
          "GO_CHROMATIN_REMODELING_AT_CENTROMERE",
          "GO_RESPONSE_TO_INTERLEUKIN_7",
          "GO_HISTONE_EXCHANGE",
          "GO_GENE_SILENCING_BY_RNA",
          "GO_CENTROMERE_COMPLEX_ASSEMBLY",
          "GO_TELOMERE_CAPPING",
          "GO_DNA_REPLICATION_INDEPENDENT_NUCLEOSOME_ORGANIZATION",
          "GO_REGULATION_OF_GENE_EXPRESSION_EPIGENETIC",
          "GO_ATP_DEPENDENT_CHROMATIN_REMODELING",
          "GO_PROTEIN_CONTAINING_COMPLEX_ASSEMBLY",
          "GO_STRESS_RESPONSE_TO_METAL_ION",
          "GO_RHOMBOMERE_DEVELOPMENT",
          "GO_RESPONSE_TO_COPPER_ION")
          
          
          
GOmolec<- c("GO_ZINC_ION_BINDING",
          "GO_TRANSITION_METAL_ION_BINDING",
          "GO_CYTIDINE_DEAMINASE_ACTIVITY",
          "GO_MRNA_3_UTR_BINDING",
          "GO_NUCLEOSOMAL_DNA_BINDING",
          "GO_PROTEIN_HETERODIMERIZATION_ACTIVITY",
          "GO_NUCLEOSOME_BINDING",
          "GO_CHROMATIN_DNA_BINDING",
          "GO_PROTEIN_DIMERIZATION_ACTIVITY")
          
GOcell<- c("GO_DNA_PACKAGING_COMPLEX",
          "GO_PROTEIN_DNA_COMPLEX",
          "GO_NUCLEAR_NUCLEOSOME",
          "GO_CHROMATIN",
          "GO_CHROMATIN",
          "GO_CHROMOSOME",
          "GO_NUCLEAR_CHROMOSOME")

#names<- tolower(names)
#names<- data.frame(names)

GObio<- tolower(GObio)
GOmolec<- tolower(GOmolec)
GOcell<- tolower(GOcell)


####### 
# alert
alert<- fread("/Users/claire/Desktop//sleepGWAS/FUMA_out/alertness/magma.gsa.out", header=T, data.table=F)

alert<- alert %>% select(FULL_NAME, P) %>% 
  mutate(set= str_split(alert$FULL_NAME,":", simplify=T)[,2],
       trait='Alertness')

# effic
effic<- fread("/Users/claire/Desktop//sleepGWAS/FUMA_out/effic/magma.gsa.out", header=T, data.table=F)

effic<- effic %>% select(FULL_NAME, P) %>%
  mutate(set= str_split(effic$FULL_NAME,":", simplify=T)[,2],
         trait='Efficiency')
head(effic)

# dur
dur<- fread("/Users/claire/Desktop//sleepGWAS/FUMA_out/duration//magma.gsa.out", header=T, data.table=F)

dur<- dur %>% select(FULL_NAME, P) %>%
  mutate(set= str_split(dur$FULL_NAME,":", simplify=T)[,2],
         trait='Duration')

#circ
cir<- fread("/Users/claire/Desktop//sleepGWAS/FUMA_out/circadian///magma.gsa.out", header=T, data.table=F)

cir<- cir %>% select(FULL_NAME, P) %>%
  mutate(set= str_split(cir$FULL_NAME,":", simplify=T)[,2],
         trait='Circadian Preference')


#insom
insom<- fread("/Users/claire/Desktop/sleepGWAS/FUMA_out/insom////magma.gsa.out", header=T, data.table=F)

insom<- insom %>% select(FULL_NAME, P) %>%
  mutate(set= str_split(insom$FULL_NAME,":", simplify=T)[,2],
         trait='Non-Insomnia')

# all these dataframes should have the same number of row bc we have not filtered


# select just the significant sets for each trait
insomsig<- insom %>% filter(P<.05) %>%
  select(set)

cirsig<- cir %>% filter(P<.05) %>%
  select(set)

alertsig<- alert %>% filter(P<.05) %>%
  select(set)

dursig<- dur %>% filter(P<.05) %>%
  select(set)

efficsig<- effic %>% filter(P<.05) %>%
  select(set)

names<- rbind(insomsig, cirsig, alertsig, dursig, efficsig)
# this has all the gene sets that are significant A LOT. now we can merge with the list of GO terms

names$GO[names$set %in% GObio]<- "GO Biological Process"
names$GO[names$set %in% GOmolec]<- "GO Molecular Process"
names$GO[names$set %in% GOcell]<- "GO Cellular Process"

names<- names %>% filter(GO=="GO Biological Process" | GO== "GO Molecular Process" | GO== "GO Cellular Process")
nrow(names) # 26 to plot on. now get all 26 for each factor

#### now need to get all together


insom<- merge(insom,names,by="set", all.y=T)
alert<- merge(alert,names,by="set", all.y=T)
dur<- merge(dur,names,by="set",all.y=T)
cir<- merge(cir,names,by="set", all.y=T)
effic<- merge(effic,names,by="set", all.y=T)

# should be 26 for all


d<- rbind(alert, effic, dur, insom,cir)
nrow(d) # should be 26*5

#BUT have to remove duplicates (if one was sig on their own and another factor was)

d<- d %>% group_by(set, trait) %>%
  slice(1) %>%
  arrange(desc(GO)) %>%
  arrange(desc(trait))

d$trait <- factor(d$trait, levels = c("Non-Insomnia", "Duration", "Efficiency", "Alertness", "Circadian Preference"))


d$`-log10p`<- -log10(d$P)

d$`GO set` <- factor(d$set, levels = d$set[1:16])


p<- ggplot(d, aes(y=`GO set`, x=`-log10p`, fill=trait)) + 
  geom_bar(stat = "identity", position="dodge", width=0.75) +
  ylab("GO term")+
  theme(text = element_text(size = 10)) +
  theme(legend.title = element_blank(),
        panel.border = element_blank(),
      panel.grid.major = element_blank(),
      panel.grid.minor = element_blank()) +
  geom_vline(xintercept = 1.301, linetype="dotted", 
                color = "red", size=1) +
  annotate(geom="text", y=12, x=3, label="GO \nBiological \nTerms", size=5)+
  annotate(geom="text", y=6, x=3, label="GO \nCellular \nTerms", size=5)+
  annotate(geom="text", y=2, x=3, label="GO \nMolecular \nTerms", size=5)+
  annotate(geom="rect", ymin=7.5, ymax=Inf, xmin=0, xmax=Inf, fill="darkblue", alpha=0.1)+
  annotate(geom="rect", ymin=4.6, ymax=7.4, xmin=0, xmax=Inf, fill="yellow", alpha=0.1)+
  annotate(geom="rect", ymin=0, ymax=4.5, xmin=0, xmax=Inf, fill="darkblue", alpha=0.1)+
  scale_fill_manual(values = fact_colors)
  
p<- p + guides(fill = guide_legend(reverse = TRUE))

ggsave(plot = p,filename = "/Users/claire/Desktop//sleepGWAS/figs/GOterms.png", width = 9, height = 7, units = "in")


head(d)

d %>% filter(P<.05) %>%
  group_by(trait) %>%
  count(trait) 

d %>% filter(P<.05) %>%
  group_by(`GO set`) %>%
  count(`GO set`) %>% 
  filter(n>1)

length(unique(d$`GO set`)) # 7/16 were associated with > 1 
```

### manhattan plots from kristen 



```{r}
library('ggplot2')

calc_limits <- function(values,fix_upper=NULL,fix_lower=NULL,lowlim_min=NULL,lowlim_max=NULL,uplim_min=NULL,uplim_max=NULL,make_symmetric=F) {
  # Check for silliness
  if(any(c(lowlim_min >= uplim_max,lowlim_min >= fix_upper,fix_lower >= uplim_max))) {
    print(cbind(fix_lower,lowlim_min,fix_upper,uplim_max))
    stop('ERROR: lowlim_min or fix_lower is >= fix_upper or uplim_max')
  }
  # Find highest and lowest values from the data
  # Also apply the "outer limits" of lowlim_min and uplim_max
  min_val <- max(c(lowlim_min,min(values,na.rm=T)),na.rm=T)
  max_val <- min(c(uplim_max,max(values,na.rm=T)),na.rm=T)
  # If symmetric, expand the other min/max accordingly
  if(make_symmetric) {
    if(max_val >= 0) {
      min_val <- min(c(min_val,-1*max_val))
      max_val <- max(c(-1*min_val,max_val))
    } else {
      stop('ERROR: Behavior of make_symmetric is undefined when the maximum value is < 0.')
    }
  }
  # Make sure the lower is at least as low as lowlim_max and the upper is at least as high as uplim_min
  min_val <- min(c(min_val,lowlim_max))
  max_val <- max(c(max_val,uplim_min))
  # Use calculated values or user-specified fixed values?
  if(is.null(fix_lower)) {
    lower <- min_val
  } else {
    lower <- fix_lower
    if(make_symmetric) { print('WARNING: Using provided fixed value for lower limit, ignoring make_symmetric.')}
    if(!is.null(lowlim_min) | !is.null(lowlim_max)) { print('WARNING: Using provided fixed value for lower limit, ignoring lowlim_min and lowlim_max.')}
  }
  if(is.null(fix_upper)) {
    upper <- max_val
  } else {
    upper <- fix_upper
    if(make_symmetric) { print('WARNING: Using provided fixed value for upper limit, ignoring make_symmetric.')}
    if(!is.null(uplim_min) | !is.null(uplim_max)) { print('WARNING: Using provided fixed value for upper limit, ignoring uplim_min and uplim_max.')}
  }
  return(c(lower,upper))
}

basic_ggplot <- function(snpdat,stat,title=NULL,xlab=NULL,ylab=NULL,space_between=0.01,
                         dotsize=1.5,legend_title=NULL,legend_position='right',basecolors=c("gray20","gray60"),chr_borders=F,
                         border_color="gray75",colorby=NULL,firstcat_nocolor=!make_color_factor,colors=NULL,
                         y_upper=NULL,y_lower=NULL,y_lowlim_min=NULL,y_lowlim_max=NULL,y_uplim_min=NULL,
                         y_uplim_max=NULL,y_symmetric=F,colors_on_top=F,contin_colorscale=NULL,make_color_factor=T,drop_unused_colors=T) {
  # Make CHR an ordered factor if it isn't already one
  if(is.numeric(snpdat$CHR)) {
    chrnum <- sort(unique(snpdat$CHR))
    snpdat$CHR <- factor(snpdat$CHR,levels=chrnum,labels=as.character(chrnum),ordered=T)
  } else if(!is.factor(snpdat$CHR) | !is.ordered(snpdat$CHR)) {
    stop('ERROR: CHR column must either be numeric or an ordered factor.')
  }

  # Make it so the minimum position on each chromosome is 1
  min_pos <- tapply(snpdat$POS,snpdat$CHR,min)
  snpdat$POS <- snpdat$POS - min_pos[snpdat$CHR] + 1

  # Get chromosome lengths, and pad for space between (if any)
  chr_length <- tapply(snpdat$POS,snpdat$CHR,max)
  if(!is.null(space_between)) {
    if(space_between < 1) { # Space given as percent of total genome length
      total_length <- sum(chr_length)
      space_percent <- space_between
      space_between <- round(space_percent*total_length)
    }
    chr_length <- chr_length + space_between
  }

  # Create a x-axis value based on cumulative chromosome lengths
  pos_shift <- head(c(0,cumsum(chr_length)),-1) # How far to shift starting POS for each chromosome to account for those before it
  names(pos_shift) <- levels(snpdat$CHR)
  snpdat$POS <- snpdat$POS + pos_shift[snpdat$CHR]

  # Make a vector of chromosome start/mid/endpoints, for borders and X-axis labels
  startpoints <- tapply(snpdat$POS,snpdat$CHR,min)
  endpoints <- tapply(snpdat$POS,snpdat$CHR,max)
  midpoints <- startpoints + (endpoints - startpoints)/2

  # Remove hidden SNPs (SNPs used for chrosome size/length but not plotted)
  if(any(colnames(snpdat)=='hide')) {
    snpdat <- snpdat[!snpdat$hide,]
  }

  # Set up for alternating chromosome colors
  # Done this way in case X/Y are included
  chr_names <- levels(snpdat$CHR)
  odd_chr <- chr_names[seq(1,length(chr_names),2)]
  snpdat$point_type <- 'Even'
  snpdat$point_type[snpdat$CHR %in% odd_chr] <- 'Odd'

  # Set up SNP colors
  if(!is.null(colorby)) {
    snpdat$point_type[!is.na(snpdat[,colorby])] <- 'Color'
    color_categorical <- is.factor(snpdat[,colorby]) | is.character(snpdat[,colorby]) | !is.null(colors)
    if(color_categorical) {
      # Set up matching of colors to categories - this will control the order they appear in the legend
      if(!is.null(colors) & !is.null(names(colors))) { # Names already supplied, keep them in the order given
        colorcats <- names(colors)
      } else {
        if(is.factor(snpdat[,colorby])) {
          colorcats <- levels(snpdat[,colorby])
        } else { # Character, go alphabetical
          colorcats <- sort(unique(snpdat[!is.na(snpdat[,colorby]),colorby]))
        }
      }
      if(!is.null(colors) & length(colorcats)!=length(colors)) {
        stop('Number of (unnamed) colors supplied does not match number of categories in data.')
      }
      # Set up colors if not supplied
      if(is.null(colors)) {
        if(length(colorcats) < 6) {
          colors <- rainbow(length(colorcats))
        } else {
          colors <- rep_len(rainbow(6),length(colorcats))
        }
      }
    }
  }

  # Make the plot
  plt <- ggplot(snpdat, aes_string(x='POS', y=stat, color=colorby)) +
    geom_point(data=snpdat[snpdat$point_type=='Odd',],color=basecolors[1],alpha=0.8, size=dotsize) +
    geom_point(data=snpdat[snpdat$point_type=='Even',],color=basecolors[2],alpha=0.8, size=dotsize) +
    scale_x_continuous(label = names(midpoints), breaks=midpoints) +
    ggtitle(title) + xlab(xlab) + ylab(ylab) + theme_bw() +
    theme(plot.title = element_text(hjust = 0.5),
      panel.border = element_blank(),
      panel.grid.major = element_blank(),
      panel.grid.minor = element_blank(),
      axis.line = element_line(linetype="solid",color="gray80"),
      axis.text.x = element_text(size=8),
      legend.position = legend_position
    )

  # Add a layer for colored points, if applicable
  if(!is.null(colorby)) {
    plt <- plt + geom_point(data=snpdat[snpdat$point_type=='Color',], alpha=1, size=dotsize)
    # Set color scales
    if(color_categorical) {
      plt <- plt + scale_color_manual(values = colors,name=legend_title,breaks=colorcats)
    } else if(!color_categorical & !is.null(contin_colorscale)) {
      plt <- plt + contin_colorscale
    }
  }

  # Add borders if requested
  if(chr_borders) {
    start_borders <- tail(startpoints,-1) # No border to left of first
    end_borders <- head(endpoints,-1) # No border to right of last
    mid_borders <- (start_borders + end_borders)/2
    borders <- data.frame(mid_borders=mid_borders)
    plt <- plt + geom_vline(aes(xintercept=mid_borders),data=borders,linetype="dashed",color=border_color)
  }

  # Add Y-limits if requested
  if(!(is.null(y_upper) & is.null(y_lower) & is.null(y_lowlim_min) & is.null(y_lowlim_max) & is.null(y_uplim_min) & is.null(y_uplim_max) & y_symmetric == F)) { # If any option relating to the Y-axis limits has been changed from its default
    ylimits <- calc_limits(values=snpdat[!snpdat$hide,stat],fix_upper=y_upper,fix_lower=y_lower,lowlim_min=y_lowlim_min,lowlim_max=y_lowlim_max,uplim_min=y_uplim_min,uplim_max=y_uplim_max,make_symmetric=y_symmetric)
    plt <- plt + coord_cartesian(ylim=ylimits)
  }
  return(plt)
}

gg_manhattan <- function(snpdat,prunesnps=T,prune_pval=0.01,prune_keep=NULL,smallest_p=NULL,pcol="P",log10p=T,sigline=5e-8,sugline=1e-6,sigline_color="black",sugline_color="gray25",y_expand=NULL,outfile=NULL,outfile_dims=c(9,5),ylab="-log10(p)",caption=NULL,subtitle=NULL,...) {
  if(!is.null(outfile)) { print(paste0("Starting ",outfile)) }
  snpdat <- snpdat[!is.na(snpdat[,pcol]),]
  if(!any(colnames(snpdat)=="hide")) { snpdat$hide <- F }
  if(prunesnps) { # Shrink the data to make the plot faster
    if(is.null(prune_keep)) { prune_keep <- rep(F,nrow(snpdat)) }
    remove <- snpdat[,pcol] >= prune_pval & !prune_keep & rep_len(c(F,rep(T,49)),nrow(snpdat))
    snpdat$hide[remove] <- T
  }
  if(!is.null(smallest_p)) {
    print(paste0("Truncating smallest p-values at ",smallest_p))
    toosig <- snpdat[,pcol] < smallest_p # Prevent extremely-significant SNPs from making the rest of the plot too small
    snpdat[toosig,pcol] <- smallest_p # Truncate at this limit
  }
  if(log10p) {
    snpdat[snpdat[,pcol]==0,pcol] <- .Machine$double.xmin # Set 0s to the smallest non-zero value that can fit in a normal numeric variable
    snpdat$log10p <- -1*log10(snpdat[,pcol])
    pcol <- 'log10p'
    if(!is.null(sigline)) { sigline <- -1*log10(sigline) }
    if(!is.null(sugline)) { sugline <- -1*log10(sugline) }
  }
  p <- basic_ggplot(snpdat,pcol,ylab=ylab,...) + theme(axis.title.y=element_text(size=rel(0.8)))
  if(!is.null(sigline)) {
    p <- p + geom_hline(yintercept=sigline,linetype="solid",color=sigline_color,size=0.25)
  }
  if(!is.null(sugline)) {
    p <- p + geom_hline(yintercept=sugline,linetype="dashed",color=sugline_color,size=0.25)
  }
  if(is.null(y_expand)) {
    p <- p + scale_y_continuous(expand = expansion(mult=c(0,0.05))) # Remove space between points and x-axis, but still allow default 5% padding at top
  } else {
    p <- p + scale_y_continuous(expand = y_expand)
  }
  if(!is.null(caption)) {
    p <- p + labs(caption=caption)
  }
  if(!is.null(subtitle)) {
    p <- p + labs(subtitle=subtitle) + theme(plot.subtitle=element_text(hjust=0.5))
  }
  # Save it or return it
  if(is.null(outfile)) {
    return(p)
  } else {
    ggsave(outfile,plot=p,width=outfile_dims[1],height=outfile_dims[2],units="in")
  }
}
```

```{r manhattan effic}
library(dplyr)
library(data.table)

# Efficiency
sumstats<- fread("/Users/claire/Desktop/sleepGWAS/sleep_rr_upd//efficiency_gsem.txt.gz", header=T, data.table = F)

sumstats<- sumstats %>% select(CHR, BP, SNP, Pval_Estimate) %>%
  rename(POS=BP, P=Pval_Estimate)

snpcolors <- c('deepskyblue2','orange')
names(snpcolors) <- c('Episodes','Efficiency') # Note that the order of the colors/names here controls the order in the legend, I can put things in non-alphabetical order this way

ep<- fread("/Users/claire/Desktop/sleepGWAS/component/episodes/IndSigSNPs.txt", header=T, data.table=F)
ep<- ep[,4]

eff<- fread("/Users/claire/Desktop/sleepGWAS/component/effic_comp/IndSigSNPs.txt", header=T, data.table=F)
eff<- eff[,4]

sumstats$snptype[sumstats$SNP %in% ep] <- 'Episodes'
sumstats$snptype[sumstats$SNP %in% eff] <- 'Efficiency'

gg_manhattan(sumstats,colorby='snptype',colors=snpcolors,prune_keep=!is.na(sumstats$snptype),outfile='/Users/claire/Desktop/sleepGWAS/figs//Effic.png',title="Efficiency Factor with Constituent GWAS'",legend_position='bottom')

p_effic<- gg_manhattan(sumstats,colorby='snptype',colors=snpcolors,prune_keep=!is.na(sumstats$snptype),title="Efficiency Factor with Constituent GWAS'",legend_position='bottom')

```

```{r}
describeBy(abcd$weekend_dur_mcq_wave_2, abcd$zyg)
```


```{r manhattan alertness}
# alertness
sumstats<- fread("/Users/claire/Desktop/sleepGWAS/sleep_rr_upd///alertness_gsem.txt.gz", header=T, data.table = F)

sumstats<- sumstats %>% select(CHR, BP, SNP, Pval_Estimate) %>%
  rename(POS=BP, P=Pval_Estimate)

snpcolors <- c('blueviolet','aquamarine', 'deeppink2')
names(snpcolors) <- c('Sleepiness','Diurnal', 'Napping') # Note that the order of the colors/names here controls the order in the legend, I can put things in non-alphabetical order this way

sl<- fread("/Users/claire/Desktop/sleepGWAS/component/sleepiness/IndSigSNPs.txt", header=T, data.table=F)
sl<- sl[,4]

diurn<- fread("/Users/claire/Desktop/sleepGWAS/component/diurnal//IndSigSNPs.txt", header=T, data.table=F)
diurn<- diurn[,4]

nap<- fread("/Users/claire/Desktop/sleepGWAS/component/napping///IndSigSNPs.txt", header=T, data.table=F)
nap<- nap[,4]

sumstats$snptype[sumstats$SNP %in% sl] <- 'Sleepiness'
sumstats$snptype[sumstats$SNP %in% diurn] <- 'Diurnal'
sumstats$snptype[sumstats$SNP %in% nap] <- 'Napping'

gg_manhattan(sumstats,colorby='snptype',colors=snpcolors,prune_keep=!is.na(sumstats$snptype),outfile='/Users/claire/Desktop/sleepGWAS/figs/Alert.png',title="Alertness Factor with Constituent GWAS'",legend_position='bottom')

p_alert<- gg_manhattan(sumstats,colorby='snptype',colors=snpcolors,prune_keep=!is.na(sumstats$snptype),title="Alertness Factor with Constituent GWAS'",legend_position='bottom')

```



```{r manhattan circadian}
# circadian
sumstats<- fread("/Users/claire/Desktop/sleepGWAS/sleep_rr_upd//circadian_pref_gsem.txt.gz", header=T, data.table = F)

sumstats<- sumstats %>% select(CHR, BP, SNP, Pval_Estimate) %>%
  rename(POS=BP, P=Pval_Estimate)

snpcolors <- c('gold2','dodgerblue1', 'tomato1', 'thistle')
names(snpcolors) <- c('Least Active 5 Hours','Most Active 10 Hours', 'Sleep Mid Point', 'Chronotype') # Note that the order of the colors/names here controls the order in the legend, I can put things in non-alphabetical order this way

l5<- fread("/Users/claire/Desktop/sleepGWAS/component/l5/IndSigSNPs.txt", header=T, data.table=F)
l5<- l5[,4]

m10<- fread("/Users/claire/Desktop/sleepGWAS/component/m10//IndSigSNPs.txt", header=T, data.table=F)
m10<- m10[,4]

mid<- fread("/Users/claire/Desktop/sleepGWAS/component/mid/IndSigSNPs.txt", header=T, data.table=F)
mid<- mid[,4]

chron<- fread("/Users/claire/Desktop/sleepGWAS/component/chrono_comp///IndSigSNPs.txt", header=T, data.table=F)
chron<- chron[,4]

sumstats$snptype[sumstats$SNP %in% l5] <- 'Least Active 5 Hours'
sumstats$snptype[sumstats$SNP %in% m10] <- 'Most Active 10 Hours'
sumstats$snptype[sumstats$SNP %in% mid] <- 'Sleep Mid Point'
sumstats$snptype[sumstats$SNP %in% chron] <- 'Chronotype'

gg_manhattan(sumstats,colorby='snptype',colors=snpcolors,prune_keep=!is.na(sumstats$snptype),outfile='/Users/claire/Desktop/sleepGWAS/figs/Chron.png',title="Circadian Preference Factor with Constituent GWAS'",legend_position='bottom')

p_circ<- gg_manhattan(sumstats,colorby='snptype',colors=snpcolors,prune_keep=!is.na(sumstats$snptype),title="Circadian Preference Factor with Constituent GWAS'",legend_position='bottom')

```

### add bar plots of number of significant SNPs per factor, indicator, overlapping and Q

```{r}

colors_manhattan<- c("Alertness Factor"= "darkgrey", "Circadian Preference Factor"="darkgrey", "Efficiency Factor"= "darkgrey", 'Least Active 5 Hours'="gold2", "Most Active 10 Hours"="dodgerblue1", "Sleep Mid Point"= "tomato1", "Chronotype"="thistle", "Efficiency"= "orange", Episodes="deepskyblue2", "Sleepiness"="blueviolet", "Diurnal"="aquamarine", Napping="deeppink2")


gsem<- fread("/Users/claire/Desktop/sleepGWAS/GSEM_GWAS_summary.csv", header=T, data.table=F)


gsem<- gsem %>% mutate(trait=fct_relevel(trait, unique(gsem$trait)))

factors<- unique(gsem$Factor)

plot_list_bar_gsem = list()
for (i in factors) {
  dat<- gsem %>% filter(Factor==i) %>%
  mutate(category=fct_relevel(category, "Significant Loci", "Shared with Factor", "Novel", "Q-SNP"))
  p<- ggplot(dat, aes(y=n_loci, x=category, fill=trait)) + 
  geom_col(position = position_dodge2(width = 0.9, preserve = "single")) +
  theme(text = element_text(size = 10)) +
  ylab("Number of Significant Loci") +
    xlab("")+
    ylim(c(0,200))+
  ggtitle(paste("Loci Associated with", i, "Factor and Indicator GWAS'",sep=" "))+
  theme(legend.title = element_blank(),
        panel.border = element_blank(),
      panel.grid.major = element_blank(),
      panel.grid.minor = element_blank(),
      plot.title = element_text(hjust = 0.5)) +
    scale_fill_manual(values = colors_manhattan)
  plot_list_bar_gsem[[i]] = p
  #ggsave(plot = p.2,filename = paste0("/Users/claire/Desktop/Desktop/sleepGWAS/figs/TWAS_bar_", i, ".png"), width = 9, height = 9, units = "in")
}


```

### final mahattan figure

**re do at somepoint for order**

```{r}
p<- ggarrange(p_circ, plot_list_bar_gsem[[1]],
             p_effic, plot_list_bar_gsem[[3]],
             p_alert, plot_list_bar_gsem[[2]], ncol = 2, nrow = 3)

ggsave(plot = p,filename = "/Users/claire/Desktop/sleepGWAS/figs/final_gsem_manhattan.PNG", width = 15, height = 15, units = "in")

```


# MASSIVE ldsc

filtered the list of phenotypes to more meaningful phenotypes (181)
add in cEF, RT, EA

```{r}

phen<- read.csv("/Users/claire/Desktop/sleepGWAS/phenotypes_massive.csv", header=T)
nrow(phen)
head(phen)


### new bonf = 181*3

bonF<- .05/(nrow(phen)*6)

# circ
circ_mass<- fread("/Users/claire/Desktop/sleepGWAS/CTG-VL/MASSIVE/circadian_ldsc.csv", header=T, data.table=F)
circ_mass$p<- as.numeric(circ_mass$p)


circ_mass<- circ_mass %>% right_join(phen) %>%
  filter(p<bonF) %>%
  mutate(factor="Circadian Preference") %>%
  select(rg, se, p, phenotype, factor, category)
nrow(circ_mass) # 17 sig associations


# eff
eff_mass<- fread("/Users/claire/Desktop/sleepGWAS/CTG-VL/MASSIVE/efficiency_ldsc.csv", header=T, data.table=F)
eff_mass$p<- as.numeric(eff_mass$p)

eff_mass<- eff_mass %>% right_join(phen) %>%
  filter(p<bonF)%>%
  mutate(factor="Efficiency") %>%
  select(rg, se, p, phenotype, factor, category) 
nrow(eff_mass) # 7 sig associations

# alert
alert_mass<- fread("/Users/claire/Desktop//sleepGWAS/CTG-VL/MASSIVE/alertness_ldsc.csv", header=T, data.table=F)
alert_mass$p<- as.numeric(alert_mass$p)

alert_mass<- alert_mass %>% right_join(phen) %>%
  filter(p<bonF) %>%
  mutate(factor="Alertness") %>%
  select(rg, se, p, phenotype, factor, category)
nrow(alert_mass)# 55 sig associations

# dur
dur_mass<- fread("/Users/claire/Desktop/sleepGWAS/CTG-VL/MASSIVE/duration_ldsc.csv", header=T, data.table=F)
dur_mass$p<- as.numeric(dur_mass$p)

dur_mass<- dur_mass %>% right_join(phen) %>%
  filter(p<bonF) %>%
  mutate(factor="Duration") %>%
  select(rg, se, p, phenotype, factor, category) 
nrow(dur_mass)# 23 sig associations

# var
var_mass<- fread("/Users/claire/Desktop/sleepGWAS/CTG-VL/MASSIVE/variability_ldsc.csv", header=T, data.table=F)
var_mass$p<- as.numeric(var_mass$p)

var_mass<- var_mass %>% right_join(phen) %>%
  filter(p<bonF) %>%
  mutate(factor="Regularity") %>%
  select(rg, se, p, phenotype, factor, category) 
nrow(var_mass)# 18 sig associations

# insom
insom_mass<- fread("/Users/claire/Desktop//sleepGWAS/CTG-VL/MASSIVE/insomnia_ldsc.csv", header=T, data.table=F)
insom_mass$p<- as.numeric(insom_mass$p)

insom_mass<- insom_mass %>% right_join(phen) %>%
  filter(p<bonF) %>%
  mutate(factor="Non-Insomnia") %>%
  select(rg, se, p, phenotype, factor, category)
nrow(insom_mass)# 68 sig associations

# was only plotting phenos with more than 1 sig factor but let's just plot all sig


# prob want to write this and edit names
mass<- rbind(circ_mass, eff_mass, alert_mass, dur_mass, insom_mass, var_mass)

## the only ones that are really way too long are in psychiatric

# maybe get rid of "Non-cancer illness code, " and "Mental health problems ever diagnosed by a professional: "


mass$phenotype<- sub(".*Non-cancer illness code, ", "", mass$phenotype)
mass$phenotype<- sub(".*Mental health problems ever ", "", mass$phenotype)
mass$phenotype<- sub(".*self-reported: ", "", mass$phenotype)
mass$phenotype<- sub(".*diagnosed by a professional: ", "", mass$phenotype)
mass$phenotype<- sub(".*Non-cancer illness code, ", "", mass$phenotype)
mass$phenotype<- sub(".*Mental health problems ever ", "", mass$phenotype)
mass$phenotype<- sub(".*self-reported: ", "", mass$phenotype)
mass$phenotype<- sub(".*diagnosed by a professional: ", "", mass$phenotype)
mass$phenotype<- sub(".*Diagnoses - main ICD10: ", "", mass$phenotype)
mass$phenotype[mass$phenotype=="Activities undertaken to treat anxiety: Talking therapies, such as psychotherapy, counselling, group therapy or CBT"]<- "Talking therapy to treat depression"
mass$phenotype[mass$phenotype=="Activities undertaken to treat anxiety: Talking therapies, such as psychotherapy, counselling, group therapy or CBT"]<- "Talking therapy to treat anxiety"
mass$phenotype[mass$phenotype=="Activities undertaken to treat depression: Talking therapies, such as psychotherapy, counselling, group therapy or CBT"]<- "Talking therapy to treat depression"
mass$phenotype[mass$phenotype=="Activities undertaken to treat depression: Talking therapies, such as psychotherapy, counselling, group therapy or CBT"]<- "Talking therapy to treat depression"
mass$phenotype[mass$phenotype=="Seen doctor (GP) for nerves, anxiety, tension or depression"]<- "Seen doctor (GP) for anxiety or depression"
mass$phenotype[mass$phenotype=="Seen a psychiatrist for nerves, anxiety, tension or depression"]<- "Seen psychiatrist for anxiety or depression"


# how many overlap? how many traits in general

head(circ_mass)

c<- circ_mass %>% mutate(circ="1") %>%
  select(phenotype, circ)
d<- dur_mass %>% mutate(dur="1") %>%
  select(phenotype, dur)
a<- alert_mass %>% mutate(alert="1") %>%
  select(phenotype, alert)
e<- eff_mass %>% mutate(eff="1") %>%
  select(phenotype, eff)
i<- insom_mass %>% mutate(insom="1") %>%
  select(phenotype, insom)
r<- var_mass %>% mutate(reg="1") %>%
  select(phenotype, reg)

t<- merge(c,d,by="phenotype", all=T)
t<- merge(t,a,by="phenotype", all=T)
t<- merge(t,e,by="phenotype", all=T)
t<- merge(t,i,by="phenotype", all=T)
t<- merge(t,r,by="phenotype", all=T)
head(t)

t$circ<- as.numeric(t$circ)
t$alert<- as.numeric(t$alert)
t$eff<- as.numeric(t$eff)
t$dur<- as.numeric(t$dur)
t$insom<- as.numeric(t$insom)
t$reg<- as.numeric(t$reg)

t<- t %>% replace(is.na(.), 0) %>%
   mutate(sum = rowSums(across(where(is.numeric)))) %>%
  group_by(phenotype) %>%
  slice(1)

table(t$sum);nrow(t)

t %>% filter(sum==5)


psych<- mass %>% filter(category=="Psychiatric")
unique(psych$phenotype)

health<- mass %>% filter(category=="Health")
unique(health$phenotype)
```

# plot LDSC

```{r}
mass$factor <- factor(mass$factor, levels = c("Circadian Preference", "Alertness", "Efficiency", "Duration", "Non-Insomnia", "Regularity"))



mass<- mass %>%
  mutate(phenotype=fct_relevel(phenotype,unique(mass$phenotype)[c(1:21, 23:33,22,34:78)]))


mass$rg<- as.numeric(mass$rg)
mass$se<- as.numeric(mass$se)
mass$CI<- mass$se*(1.96)

# mass$phenotype <- str_wrap(mass$phenotype, width = 15)

### plot 1 

cats<- c("Psychiatric", "Health", "Lifecourse", "Environmental", "Sleep","Education", "Substance Use", "Activity", "Anthropometric", "Blood Assay")
mass_plots<- list()

for (i in cats) {
  dat<- subset(mass, mass$category==i)
  p<- ggplot(dat, aes(x=rg, y=phenotype, colour=factor, shape=factor, xmin = rg-CI, xmax = rg+CI)) +
  geom_point(size=3,position=position_dodge(width=0.5)) + 
   geom_errorbarh(aes(x = CI), height=.4, position=position_dodge(width=0.5))+
  xlab("rG") +
  ylab("Phenotype") +
  scale_y_discrete(labels = function(x) str_wrap(x, width = 50))+
  ggtitle(i)+
  geom_vline(xintercept = 0, linetype="dashed")+
  theme(legend.title = element_blank(),
        panel.border = element_blank(),
      panel.grid.major = element_blank(),
      panel.grid.minor = element_blank(),
      plot.title = element_text(hjust = 0.5, size = 17),
      legend.position="bottom")+
  scale_color_manual(values = fact_colors) 
  mass_plots[[i]]<- p
}


t1<- ggarrange(mass_plots[[1]], mass_plots[[2]],
             mass_plots[[3]], mass_plots[[4]],
             mass_plots[[5]],
             ncol = 1, nrow = 5,common.legend = F)



t2<- ggarrange(mass_plots[[6]], mass_plots[[7]],
             mass_plots[[8]], mass_plots[[9]],
             mass_plots[[10]],
             ncol = 1, nrow = 5,common.legend = F)


#t2<- annotate_figure(t1, top = text_grob("Batch Genetic Correlations between Sleep Health and External Traits", size = 14))

# ggsave(plot = t1,filename = "/Users/claire/Desktop//Desktop/sleepGWAS/figs//LDSC_else.png", width = 7, height = 24, units = "in")

ldsc<- ggarrange(t1, t2, ncol=2, common.legend = F)

ggsave(plot = ldsc,filename = "/Users/claire/Desktop/sleepGWAS/figs//LDSC.png", width = 14, height = 24, units = "in")

mass %>% distinct(phenotype, category) %>%
  group_by(category) %>%
  count(category) %>%
  slice(1)



#### edit for defense:
# plotting life course and env together

mass$category[mass$category=="Lifecourse"]<- "Environmental"

dat<- mass %>% filter(category=="Environmental")

dat$category<- "Environment & Lifecourse"

ggplot(dat, aes(x=rg, y=phenotype, colour=factor, shape=factor, xmin = rg-CI, xmax = rg+CI)) +
  geom_point(size=3,position=position_dodge(width=0.5)) + 
   geom_errorbarh(aes(x = CI), height=.4, position=position_dodge(width=0.5))+
  xlab("rG") +
  ylab("Phenotype") +
  scale_y_discrete(labels = function(x) str_wrap(x, width = 50))+
  ggtitle("Environment & Lifecourse")+
  geom_vline(xintercept = 0, linetype="dashed")+
  theme(legend.title = element_blank(),
        panel.border = element_blank(),
      panel.grid.major = element_blank(),
      panel.grid.minor = element_blank(),
      plot.title = element_text(hjust = 0.5, size = 17),
      legend.position="bottom")+
  scale_color_manual(values = fact_colors) +
  guides(shape = "none")
```


## could alsp plot non organized for supp!

```{r}

# circ
circ_mass<- fread("/Users/claire/Desktop/sleepGWAS/CTG-VL/MASSIVE/circadian_ldsc.csv", header=T, data.table=F)


circ_mass<- circ_mass%>%
  mutate(factor="Circadian Preference") %>%
  select(rg, se, p, phenotype, factor)


# eff
eff_mass<- fread("/Users/claire/Desktop/sleepGWAS/CTG-VL/MASSIVE/efficiency_ldsc.csv", header=T, data.table=F)

eff_mass<- eff_mass %>% 
  mutate(factor="Efficiency") %>%
  select(rg, se, p, phenotype, factor) 

# alert
alert_mass<- fread("/Users/claire/Desktop//sleepGWAS/CTG-VL/MASSIVE/alertness_ldsc.csv", header=T, data.table=F)

alert_mass<- alert_mass %>% 
  mutate(factor="Alertness") %>%
  select(rg, se, p, phenotype, factor)

# dur
dur_mass<- fread("/Users/claire/Desktop/sleepGWAS/CTG-VL/MASSIVE/duration_ldsc.csv", header=T, data.table=F)

dur_mass<- dur_mass %>% 
  mutate(factor="Duration") %>%
  select(rg, se, p, phenotype, factor) 

# var
var_mass<- fread("/Users/claire/Desktop/sleepGWAS/CTG-VL/MASSIVE/variability_ldsc.csv", header=T, data.table=F)

var_mass<- var_mass %>% 
  mutate(factor="Regularity") %>%
  select(rg, se, p, phenotype, factor) 

# insom
insom_mass<- fread("/Users/claire/Desktop//sleepGWAS/CTG-VL/MASSIVE/insomnia_ldsc.csv", header=T, data.table=F)

insom_mass<- insom_mass %>%
  mutate(factor="Non-Insomnia") %>%
  select(rg, se, p, phenotype, factor)

# was only plotting phenos with more than 1 sig factor but let's just plot all sig


# prob want to write this and edit names
mass<- rbind(circ_mass, eff_mass, alert_mass, dur_mass, insom_mass, var_mass)

### new bonf = 192*3

bonF<- .05/(nrow(mass)*6)

mass$se<- as.numeric(mass$se)
mass$p<- as.numeric(mass$p)
mass$rg<- as.numeric(mass$rg)

mass<- mass %>% filter(p<bonF) %>%
  mutate(CI=1.96*se)

nrow(mass)
table(mass$factor)

mass$factor <- factor(mass$factor, levels = c("Circadian Preference", "Alertness", "Efficiency", "Duration", "Non-Insomnia", "Regularity"))


mass$phenotype<- sub(".*Non-cancer illness code, ", "", mass$phenotype)
mass$phenotype<- sub(".*Mental health problems ever ", "", mass$phenotype)
mass$phenotype<- sub(".*self-reported: ", "", mass$phenotype)
mass$phenotype<- sub(".*diagnosed by a professional: ", "", mass$phenotype)
mass$phenotype<- sub(".*Diagnoses - main ICD10: ", "", mass$phenotype)
mass$phenotype[mass$phenotype=="Activities undertaken to treat anxiety: Talking therapies, such as psychotherapy, counselling, group therapy or CBT"]<- "Talking therapy to treat depression"
mass$phenotype[mass$phenotype=="Activities undertaken to treat anxiety: Talking therapies, such as psychotherapy, counselling, group therapy or CBT"]<- "Talking therapy to treat anxiety"
mass$phenotype[mass$phenotype=="Activities undertaken to treat depression: Talking therapies, such as psychotherapy, counselling, group therapy or CBT"]<- "Talking therapy to treat depression"
mass$phenotype[mass$phenotype=="Activities undertaken to treat depression: Talking therapies, such as psychotherapy, counselling, group therapy or CBT"]<- "Talking therapy to treat depression"
mass$phenotype[mass$phenotype=="Seen doctor (GP) for nerves, anxiety, tension or depression"]<- "Seen doctor (GP) for anxiety or depression"
mass$phenotype[mass$phenotype=="Seen a psychiatrist for nerves, anxiety, tension or depression"]<- "Seen psychiatrist for anxiety or depression"

### split into several plots

mass<- mass %>% arrange(phenotype)
nums<- nrow(mass)/3
mass1<- mass[1:nums,] 

p<- ggplot(mass1, aes(x=rg, y=phenotype, colour=factor, shape=factor, xmin = rg-CI, xmax = rg+CI)) +
  geom_point(size=5,position=position_dodge(width=0.5)) + 
   geom_errorbarh(aes(x = CI), height=.4, position=position_dodge(width=0.5))+
  xlab("rG") +
  ylab("Phenotype") + 
  scale_y_discrete(labels = function(x) str_wrap(x, width = 55))+
  geom_vline(xintercept = 0, linetype="dashed")+
  theme(legend.title = element_blank(),
        panel.border = element_blank(),
      panel.grid.major = element_blank(),
      panel.grid.minor = element_blank(),
      plot.title = element_text(hjust = 0.5))+
  scale_color_manual(values = fact_colors) 

ggsave(plot = p,filename = "/Users/claire/Desktop//sleepGWAS/figs/LDSC_supp1.png", width = 9, height = 24, units = "in")

mass2<- mass[nums+1:nums+1,] ## should be 216:431

p<- ggplot(mass2, aes(x=rg, y=phenotype, colour=factor, shape=factor, xmin = rg-CI, xmax = rg+CI)) +
  geom_point(size=5,position=position_dodge(width=0.5)) + 
   geom_errorbarh(aes(x = CI), height=.4, position=position_dodge(width=0.5))+
  xlab("rG") +
  ylab("Phenotype") + 
  scale_y_discrete(labels = function(x) str_wrap(x, width = 55))+
  geom_vline(xintercept = 0, linetype="dashed")+
  theme(legend.title = element_blank(),
        panel.border = element_blank(),
      panel.grid.major = element_blank(),
      panel.grid.minor = element_blank(),
      plot.title = element_text(hjust = 0.5))+
  scale_color_manual(values = fact_colors) 

ggsave(plot = p,filename = "/Users/claire/Desktop//sleepGWAS/figs/LDSC_supp2.png", width = 9, height = 24, units = "in")

mass3<- mass[(nums*2)+2:nums,] ## should be 432:645

p<- ggplot(mass3, aes(x=rg, y=phenotype, colour=factor, shape=factor, xmin = rg-CI, xmax = rg+CI)) +
  geom_point(size=5,position=position_dodge(width=0.5)) + 
   geom_errorbarh(aes(x = CI), height=.4, position=position_dodge(width=0.5))+
  xlab("rG") +
  ylab("Phenotype") + 
  scale_y_discrete(labels = function(x) str_wrap(x, width = 55))+
  geom_vline(xintercept = 0, linetype="dashed")+
  theme(legend.title = element_blank(),
        panel.border = element_blank(),
      panel.grid.major = element_blank(),
      panel.grid.minor = element_blank(),
      plot.title = element_text(hjust = 0.5))+
  scale_color_manual(values = fact_colors) 

ggsave(plot = p,filename = "/Users/claire/Desktop/sleepGWAS/figs/LDSC_supp3.png", width = 9, height = 24, units = "in")

```


## supp table for massive

```{r}
phen<- read.csv("/Users/claire/Desktop/sleepGWAS/phenotypes_massive.csv", header=T)
nrow(phen)
head(phen)


# circ
circ_mass<- fread("/Users/claire/Desktop/sleepGWAS/CTG-VL/MASSIVE/circadian_ldsc.csv", header=T, data.table=F)
circ_mass$p<- as.numeric(circ_mass$p)

circ_mass$category <- factor(circ_mass$phenotype, levels=phen$phenotype, labels=phen$category)

circ_mass<- circ_mass %>% mutate(sleep="Circadian Preference") %>%
  select(sleep, category, phenotype, rg, se, z, p, N, NCases, NControls, Cohort, URL)



# eff
eff_mass<- fread("/Users/claire/Desktop/sleepGWAS/CTG-VL/MASSIVE/efficiency_ldsc.csv", header=T, data.table=F)
eff_mass$p<- as.numeric(eff_mass$p)

eff_mass$category <- factor(eff_mass$phenotype, levels=phen$phenotype, labels=phen$category)

eff_mass<- eff_mass %>% mutate(sleep="Efficiency") %>%
  select(sleep,category, phenotype, rg, se, z, p, N, NCases, NControls, Cohort, URL)


# alert
alert_mass<- fread("/Users/claire/Desktop//sleepGWAS/CTG-VL/MASSIVE/alertness_ldsc.csv", header=T, data.table=F)
alert_mass$p<- as.numeric(alert_mass$p)

alert_mass$category <- factor(alert_mass$phenotype, levels=phen$phenotype, labels=phen$category)

alert_mass<- alert_mass %>% mutate(sleep="Alertness") %>%
  select(sleep,category, phenotype, rg, se, z, p, N, NCases, NControls, Cohort, URL)

# dur
dur_mass<- fread("/Users/claire/Desktop/sleepGWAS/CTG-VL/MASSIVE/duration_ldsc.csv", header=T, data.table=F)
dur_mass$p<- as.numeric(dur_mass$p)

dur_mass$category <- factor(dur_mass$phenotype, levels=phen$phenotype, labels=phen$category)

dur_mass<- dur_mass %>% mutate(sleep="Duration") %>%
  select(sleep,category, phenotype, rg, se, z, p, N, NCases, NControls, Cohort, URL)

# var
var_mass<- fread("/Users/claire/Desktop/sleepGWAS/CTG-VL/MASSIVE/variability_ldsc.csv", header=T, data.table=F)
var_mass$p<- as.numeric(var_mass$p)

var_mass$category <- factor(var_mass$phenotype, levels=phen$phenotype, labels=phen$category)

var_mass<- var_mass %>% mutate(sleep="Variability") %>%
  select(sleep,category, phenotype, rg, se, z, p, N, NCases, NControls, Cohort, URL)

# insom
insom_mass<- fread("/Users/claire/Desktop//sleepGWAS/CTG-VL/MASSIVE/insomnia_ldsc.csv", header=T, data.table=F)
insom_mass$p<- as.numeric(insom_mass$p)

insom_mass$category <- factor(insom_mass$phenotype, levels=phen$phenotype, labels=phen$category)

insom_mass<- insom_mass %>% mutate(sleep="Insomnia") %>%
  select(sleep, category, phenotype, rg, se, z, p, N, NCases, NControls, Cohort, URL)


sleep<- c("circ", "eff", "alert", "dur", "insom", "var")

for (i in sleep) {
  fwrite(eval(parse(text=paste(i, "mass", sep="_"))), paste0('/Users/claire/Desktop/', i, "MASS.csv"))
}

```


# twas

p val = best p val

will need to think about the order of these brain regions. potentially plot differently. maybe something like a pie chart?

```{r}

bonP<- 5e-6

# circ
circ<- fread("/Users/claire/Desktop//sleepGWAS/CTG-VL/metaxcan/circadian_metax.csv", header=T, data.table=F)
head(circ)
nrow(circ)

sig<- circ %>% filter(p_i_best<bonP)
table(sig$t_i_best)

circ<- circ %>% select(gene_name, p_i_best, t_i_best,) %>%
  mutate(p_i_best= -log10(p_i_best),
         trait="Circadian Preference") %>%
  rename(Pvalue=p_i_best, Best_Tissue=t_i_best)

# alert
alert<- fread("/Users/claire/Desktop//sleepGWAS/CTG-VL/metaxcan/alertness_metax.csv", header=T, data.table=F)
head(alert)
nrow(alert)

sig<- alert %>% filter(p_i_best<bonP)
table(sig$t_i_best)

alert<- alert %>% select(gene_name, p_i_best, t_i_best,) %>%
  mutate(p_i_best= -log10(p_i_best),
         trait="Alertness") %>%
  rename(Pvalue=p_i_best, Best_Tissue=t_i_best)

# effic
effic<- fread("/Users/claire/Desktop//sleepGWAS/CTG-VL/metaxcan/efficiency_metax.csv", header=T, data.table=F)
head(effic)
nrow(effic)

sig<- effic %>% filter(p_i_best<bonP)
table(sig$t_i_best)

effic<- effic %>% select(gene_name, p_i_best, t_i_best,) %>%
  mutate(p_i_best= -log10(p_i_best),
         trait="Efficiency") %>%
  rename(Pvalue=p_i_best, Best_Tissue=t_i_best)

# dur
dur<- fread("/Users/claire/Desktop//sleepGWAS/CTG-VL/metaxcan/duration_metax.csv", header=T, data.table=F)

sig<- dur %>% filter(p_i_best<bonP)
table(sig$t_i_best)

dur<- dur %>% select(gene_name, p_i_best, t_i_best,) %>%
  mutate(p_i_best= -log10(p_i_best),
         trait="Duration") %>%
  rename(Pvalue=p_i_best, Best_Tissue=t_i_best)

# insom
insom<- fread("/Users/claire/Desktop//sleepGWAS/CTG-VL/metaxcan/insomnia_metax.csv", header=T, data.table=F)
head(insom)
nrow(insom)

sig<- insom %>% filter(p_i_best<bonP)
table(sig$t_i_best)

insom<- insom %>% select(gene_name, p_i_best, t_i_best,) %>%
  mutate(p_i_best= -log10(p_i_best),
         trait="Non-Insomnia") %>%
  rename(Pvalue=p_i_best, Best_Tissue=t_i_best)

# var
var<- fread("/Users/claire/Desktop//sleepGWAS/CTG-VL/metaxcan/variability_metax.csv", header=T, data.table=F)
head(var)
nrow(var)

sig<- var %>% filter(p_i_best<bonP)
table(sig$t_i_best)

var<- var %>% select(gene_name, p_i_best, t_i_best,) %>%
  mutate(p_i_best= -log10(p_i_best),
         trait="Regularity") %>%
  rename(Pvalue=p_i_best, Best_Tissue=t_i_best)

```

plot TWAS

*1/20 update*

-plot the number of significant genes, not the sum of the p values
-significant at GWAS threshold
-make a figure that has the TWAS MANHATTAN side by side with the stacked bar plot of number of significant genes


```{r}
library(stringr)
library(forcats)
twas<- rbind(dur,circ, alert,effic,insom)

t<- "Brain_Caudate_basal_ganglia"

twas$tissue<- substring(twas$Best_Tissue, regexpr("_", twas$Best_Tissue) + 1, nchar(twas$Best_Tissue))


twas$trait<- factor(twas$trait, levels = c("Alertness", "Circadian Preference", "Duration", "Efficiency", "Non-Insomnia", "Regularity"))


twas<- twas %>% mutate(best_tissue=fct_relevel(tissue, "Cortex", "Frontal_Cortex_BA9", "Anterior_cingulate_cortex_BA24", "Caudate_basal_ganglia", "Putamen_basal_ganglia", "Nucleus_accumbens_basal_ganglia", "Brain_Hypothalamus", "Amygdala", "Hippocampus", "Cerebellum", "Cerebellar_Hemisphere", "Substantia_nigra", "Spinal_cord_cervical_c-1"))



sleep<- unique(twas$trait)

plot_list_manhattan<- list()
for (i in sleep) {
  dat<- twas %>% filter(trait==i)
  p<- ggplot(dat, aes(x=best_tissue, y=Pvalue, color=best_tissue)) + 
    geom_point(size=1, position = "jitter") +
  xlab("Tissue type") +
  ylab("-log10 p-value") +
    ggtitle(paste("TWAS of 13 GTEX Brain Regions for", i, sep=" "))+
  geom_hline(yintercept = 5.301, linetype="dashed")+
  theme(legend.title = element_blank(),
        panel.border = element_blank(),
      panel.grid.major = element_blank(),
          axis.ticks.x=element_blank(),
      axis.text.x=element_blank(),
      panel.grid.minor = element_blank())
  plot_list_manhattan[[i]]<- p
  #ggsave(plot = p,filename = paste0("/Users/claire/Desktop/Desktop/sleepGWAS/figs/TWAS_", i, ".png"), width = 9, height = 9, units = "in")
}



# how many significant genes per brain region per trait?

# ggsave(plot = p,filename = "/Users/clairemorrison/Desktop/TWAS.png", width = 9, height = 9, units = "in")


### playing around with new ways to plot

## try some sort of overlap plot... plotting genes that overlap > 5 times?

over5<- twas %>% count(gene_name) %>%
  filter(n>5)
nrow(over5)

# that doesn't seem to reduce it

# range of total p values per tissue

tapply(twas$Pvalue, twas$Best_Tissue, sum)

# broken down by trait
with(twas, tapply(Pvalue, list(Best_Tissue,trait), sum))

min(with(twas, tapply(Pvalue, list(Best_Tissue,trait), sum))); max(with(twas, tapply(Pvalue, list(Best_Tissue,trait), sum)))

# average log 10 p value per tissue type and trait
with(twas, tapply(Pvalue, list(Best_Tissue,trait), mean))

min(with(twas, tapply(Pvalue, list(Best_Tissue,trait), mean))); max(with(twas, tapply(Pvalue, list(Best_Tissue,trait), mean)))

# number of genes enriched per tissue by trait

for (i in sleep) {
  count<- twas %>% filter(trait==i) %>%
  count(Best_Tissue, gene_name) %>%
  group_by(Best_Tissue) %>%
  mutate(genes_per_t= sum(n)) %>%
  slice(n=1)
  print(min(count$genes_per_t))
  print(max(count$genes_per_t))
}

### same but only significant
for (i in sleep) {
  count<- twas %>% filter(trait==i & Pvalue>5.301) %>%
  count(Best_Tissue, gene_name) %>%
  group_by(Best_Tissue) %>%
  mutate(genes_per_t= sum(n)) %>%
  slice(n=1)
  print(min(count$genes_per_t))
  print(max(count$genes_per_t))
}


twas %>% filter(Pvalue>5.301) %>%
  group_by(tissue) %>%
  count(tissue) %>%
  arrange(desc(n))

print(twas %>% filter(Pvalue>5.301) %>%
  group_by(Best_Tissue, trait) %>%
  count(Best_Tissue) %>%
    arrange(desc(n)), n=75)

twas %>% filter(Pvalue>5.301) %>%
  group_by(Best_Tissue, trait) %>%
  count(Best_Tissue) %>%
  filter(trait=="Non-Insomnia")

twas %>% filter(Pvalue>5.301) %>%
  group_by(Best_Tissue, trait) %>%
  count(Best_Tissue) %>%
  filter(trait=="Duration")

twas %>% filter(Pvalue>5.301) %>%
  group_by(Best_Tissue, trait) %>%
  count(Best_Tissue) %>%
  filter(trait=="Alertness")

twas %>% filter(Pvalue>5.301) %>%
  group_by(Best_Tissue, trait) %>%
  count(Best_Tissue) %>%
  filter(trait=="Efficiency")

twas %>% filter(Pvalue>5.301) %>%
  group_by(Best_Tissue, trait) %>%
  count(Best_Tissue) %>%
  filter(trait=="Circadian Preference")

## try a bar plot --> this time only significant genes

sig<- twas %>% filter(Pvalue>5.301)

sig %>% arrange(desc(Pvalue)) %>%
  group_by(trait) %>%
  slice(n=1)


p.2<- ggplot(sig, aes(fill=trait, y=Pvalue, x=best_tissue)) + 
    geom_bar(position="stack", stat="identity")  +
  ylab("-log10 p-value") +
  xlab("Tissue type")+
  ggtitle("Aggregated genome-wide significant -log 10 p-values for all genes enriched per tissue type")+
  theme(legend.title = element_blank(),
        panel.border = element_blank(),
      panel.grid.major = element_blank(),
          axis.ticks.x=element_blank(),
      axis.text.x=element_blank(),
      panel.grid.minor = element_blank())+
  theme(axis.text.x=element_text(angle = 90, vjust = 0, hjust = 1))


## number of unique genes
twas %>% count(gene_name)


ggsave(plot = p.2,filename = "/Users/claire/Desktop/Desktop/TWAS.2.png", width = 9, height = 9, units = "in")

### plot individual bar plots per trait with NUMBER of sig genes instead. put side by side with manhattan

plot_list_bar = list()
for (i in sleep) {
  plotsum<- sig %>% filter(trait==i) %>%
  select(trait, best_tissue, gene_name) %>%
  group_by(best_tissue, trait) %>%
  count(gene_name) %>%
  mutate(genes_per_t= sum(n)) %>%
  slice(n=1)
  p<- ggplot(plotsum, aes(y=genes_per_t, x=best_tissue, fill=best_tissue)) + 
    geom_bar(stat="identity")  +
  ylab("Number of Genes") +
  xlab("Tissue type")+
    ylim(c(0,30))+
  ggtitle(paste("Number of Genome-wide Significant Genes per Tissue Type for", i, sep=" "))+
  theme(legend.title = element_blank(),
        panel.border = element_blank(),
      panel.grid.major = element_blank(),
          axis.ticks.x=element_blank(),
      axis.text.x=element_blank(),
      panel.grid.minor = element_blank())
  plot_list_bar[[i]] = p
  #ggsave(plot = p.2,filename = paste0("/Users/claire/Desktop/Desktop/sleepGWAS/figs/TWAS_bar_", i, ".png"), width = 9, height = 9, units = "in")
}

t<- ggarrange(plot_list_manhattan[[2]], plot_list_bar[[2]],
             plot_list_manhattan[[3]], plot_list_bar[[3]],
             plot_list_manhattan[[4]], plot_list_bar[[4]],
             plot_list_manhattan[[1]], plot_list_bar[[1]],
             plot_list_manhattan[[5]], plot_list_bar[[5]], ncol = 2, nrow = 5,common.legend = T)

ggsave(plot = t,filename = "/Users/claire/Desktop/Desktop/sleepGWAS/figs/TWAS_Full.PNG", width = 15, height = 20, units = "in")


```



Trying to plot as a sankey diagram
```{r sankey, eval=F}
library(networkD3)

traits<- unique(twas$trait)
tissues<- unique(twas$Best_Tissue)


dur<- data.frame(trait=rep("Duration", 13), tissue=tissues, source=rep(1,13), target=seq(7,19,1))

for (t in tissues) {
    dat<- twas %>% filter(trait=="Duration")  %>% 
    filter(Best_Tissue==t)
    dur$Value[which(dur$tissue==t)]<- sum(dat$Pvalue)      
}


circ<- data.frame(trait=rep("Circadian Preference", 13), tissue=tissues, source=rep(2,13), target=seq(7,19,1))

for (t in tissues) {
    dat<- twas %>% filter(trait=="Circadian Preference")  %>% 
    filter(Best_Tissue==t)
    circ$Value[which(circ$tissue==t)]<- sum(dat$Pvalue)      
}


alert<- data.frame(trait=rep("Alertness", 13), tissue=tissues, source=rep(3,13), target=seq(7,19,1))

for (t in tissues) {
    dat<- twas %>% filter(trait=="Alertness")  %>% 
    filter(Best_Tissue==t)
    alert$Value[which(alert$tissue==t)]<- sum(dat$Pvalue)      
}

effic<- data.frame(trait=rep("Efficiency", 13), tissue=tissues, source=rep(4,13), target=seq(7,19,1))

for (t in tissues) {
    dat<- twas %>% filter(trait=="Efficiency")  %>% 
    filter(Best_Tissue==t)
    effic$Value[which(effic$tissue==t)]<- sum(dat$Pvalue)      
}

insom<- data.frame(trait=rep("Non-Insomnia", 13), tissue=tissues, source=rep(5,13), target=seq(7,19,1))

for (t in tissues) {
    dat<- twas %>% filter(trait=="Non-Insomnia")  %>% 
    filter(Best_Tissue==t)
    insom$Value[which(insom$tissue==t)]<- sum(dat$Pvalue)      
}

reg<- data.frame(trait=rep("Regularity", 13), tissue=tissues, source=rep(6,13), target=seq(7,19,1))

for (t in tissues) {
    dat<- twas %>% filter(trait=="Regularity")  %>% 
    filter(Best_Tissue==t)
    reg$Value[which(reg$tissue==t)]<- sum(dat$Pvalue)      
}


links<- rbind(dur, circ, alert, effic, insom, reg)
links$trait<- NULL
links$tissue<- NULL

nodes<- data.frame(names=c(traits, tissues))

links$source<- links$source-1 ## needs to be zero indexed aka start at 0
links$target<- links$target-1


p<- sankeyNetwork(Links = links, Nodes = nodes, Source = "source",
              Target = "target", Value = "Value", NodeID = "names",
              units = "TWh", fontSize = 12, nodeWidth = 30)



```


# phewas
take SNPs that were "novel" in our GWAS and see what they associated with in a PHEWAS
will need to compare/ plot

<https://mrcieu.github.io/ieugwasr/reference/index.html>

read this!! 

<https://www.biorxiv.org/content/10.1101/2020.08.10.244293v1.full>
```{r}
library(ieugwasr)

### effic
effic<- fread("/Users/claire/Desktop/sleepGWAS/effic_novel_SNPs.txt", header=T, data.table=F)
effic<- effic$SNP

efficPHE<- phewas(effic, pval = 1e-05, batch = c(), access_token = check_access_token())

head(efficPHE)
table(efficPHE$rsid)

fwrite(efficPHE, "/Users/claire/Desktop/sleepGWAS/PHEWASeffic.csv", sep='\t')

## alert

alert<- fread("/Users/claire/Desktop/sleepGWAS/alert_novel_SNPs.txt", header=T, data.table=F)
alert<- alert$SNP

alertPHE<- phewas(alert, pval = 1e-05, batch = c(), access_token = check_access_token())

head(alertPHE)
table(alertPHE$rsid)

fwrite(alertPHE, "/Users/claire/Desktop/sleepGWAS/PHEWASalert.csv", sep='\t')

## circ

circ<- fread("/Users/claire/Desktop//sleepGWAS/circ_novel_SNPs.txt", header=T, data.table=F)
circ<- circ$SNP

# too many SNPs for circ, split into multiple queries
length(alert)
length(circ)

circ1<- circ[1:22]
circ2<- circ[23:43]
circ3<- circ[44:length(circ)]
length(circ1)+length(circ2)+length(circ3)==length(circ)


circPHE1<- phewas(circ1, pval = 1e-05, batch = c(), access_token = check_access_token())
circPHE2<- phewas(circ2, pval = 1e-05, batch = c(), access_token = check_access_token())
circPHE3<- phewas(circ3, pval = 1e-05, batch = c(), access_token = check_access_token())

circPHE<- rbind(circPHE1, circPHE2, circPHE3)
table(circPHE$rsid)

fwrite(circPHE, "/Users/claire/Desktop/sleepGWAS//PHEWAScirc.csv", sep='\t')


circPHE$factor<- rep("Circadian Preference", 1, nrow(circPHE))
alertPHE$factor<- rep("Alertness", 1, nrow(alertPHE))
efficPHE$factor<- rep("Efficiency", 1, nrow(efficPHE))

fullPHE<- rbind(alertPHE, efficPHE, circPHE)
head(fullPHE)

table(fullPHE$factor)

#fwrite(fullPHE, "/Users/claire/Desktop/sleepGWAS/PHEWAS.csv", sep=',')

```

classified PHEWAS results by hand
### plot PHEWAS

```{r}
library(forcats)
phewas<- fread("/Users/claire/Desktop/sleepGWAS/PHEWAS.csv", header=T, data.table=F)
head(phewas)


phewas<- phewas %>% select(rsid, beta,se,p,trait,factor,category) %>%
  filter(!is.na(category)) %>%
  mutate(log10p= -log10(p),
         ID = row_number(),
         factor= fct_relevel(factor, "Circadian Preference", "Efficiency", "Alertness"))

table(phewas$category)
table(phewas$factor)

range(phewas$beta)
range(phewas$log10p)

phewas %>% group_by(factor) %>%
  count(rsid) %>%
  mutate(snps=sum(n)) %>%
  slice(1)

phewas %>% group_by(factor) %>%
  count(category) %>%
  arrange(desc(n))

phewas %>% group_by(factor) %>%
  count(category) %>%
  arrange(desc(n)) %>% 
  filter(factor=="Alertness")

phewas %>% group_by(factor) %>%
  count(category) %>%
  arrange(desc(n)) %>% 
  filter(factor=="Circadian Preference")

phewas %>% group_by(factor) %>%
  count(category) %>%
  arrange(desc(n)) %>% 
  filter(factor=="Efficiency")

# antho number one for both
# circ = anthro, health, blood
# effic= anthro, blood, health
# alert= anthro, blood, sleep


## top 3 associations each factor

dat<- list()
for (f in unique(phewas$factor)) {
  dat[[f]]<- phewas %>% filter(factor==f) %>%
    select(rsid, beta, p, category, trait)
  for (c in unique(dat[[f]]$category)) {
    dat[[paste0(f,c)]]<- dat[[f]] %>% filter(category==c) %>%
      arrange(-desc(p)) %>%
      slice(1:3)
  }
}


```



for alertness, label top blood assay, health, psychiatric, sleep

for circadian preference, label top anthropocentric, cardiovascular, health, psychiatric

for efficiency, label top blood assay, sleep

```{r}

phewas$trait<- sub(".*Non-cancer illness code self-reported: ", "", phewas$trait)

phewas$log10p[phewas$log10p==Inf]<- 10
phewas$trait[phewas$trait=="Duration to first press of snap-button in each round"]<- "RT"
phewas$trait[phewas$trait=="Mean time to correctly identify matches"]<- "RT"
phewas$trait[phewas$trait=="Mean platelet (thrombocyte) volume"]<- "Thrombocyte volume"
phewas$trait[phewas$trait=="Non-cancer illness code  self-reported: malabsorption/coeliac disease"]<- "Coeliac disease"
phewas$trait[phewas$trait=="Impedance of whole body"]<- "Impedance \nof whole body"


Alertness_labels<- c("Blood Assay", "Health", "Reproductive", "Sleep", "Cognitive", "Cardiovascular")
Circadian_Preference_labels<- c("Anthropometric", "Health", "Reproductive", "Cardiovascular", "Blood Assay", "Education", "Sleep")
Efficiency_labels<- c("Blood Assay", "Sleep", "Health", "Anthropometric")


phewas$trait <- str_wrap(phewas$trait, width = 12)

phewas<- phewas %>% filter(category!="Unclassified")


# effic
effic_labeled<- data.frame()
effic_dat<- list()
effic<- phewas %>% filter(factor=="Efficiency") 
for (c in unique(effic$category)) {
  effic_dat[[c]]<- effic %>% filter(category==c) %>%
    arrange(desc(log10p))
  effic_dat[[c]]$label<- ifelse(rownames(effic_dat[[c]])==1 & effic_dat[[c]]$category %in% Efficiency_labels, "yes", NA)
  effic_labeled<- rbind(effic_dat[[c]], effic_labeled)
}

# circ 
circ_labeled<- data.frame()
circ_dat<- list()
circ<- phewas %>% filter(factor=="Circadian Preference") 
for (c in unique(circ$category)) {
  circ_dat[[c]]<- circ %>% filter(category==c) %>%
    arrange(desc(log10p))
  circ_dat[[c]]$label<- ifelse(rownames(circ_dat[[c]])==1 & circ_dat[[c]]$category %in% Circadian_Preference_labels, "yes", NA)
  circ_labeled<- rbind(circ_dat[[c]], circ_labeled)
}

# alert
alert_labeled<- data.frame()
alert_dat<- list()
alert<- phewas %>% filter(factor=="Alertness") 
for (c in unique(alert$category)) {
  alert_dat[[c]]<- alert %>% filter(category==c) %>%
    arrange(desc(log10p))
  alert_dat[[c]]$label<- ifelse(rownames(alert_dat[[c]])==1 & alert_dat[[c]]$category %in% Alertness_labels, "yes", NA)
  alert_labeled<- rbind(alert_dat[[c]], alert_labeled)
}


nrow(alert_labeled); nrow(circ_labeled); nrow(effic_labeled)
table(phewas$factor) ## checks out

phefull<- rbind(alert_labeled, circ_labeled, effic_labeled)


phewas_plt_list<- list()
for (i in unique(phefull$factor)) {
  dat<- phefull %>% filter(factor==i)
  plt<- ggplot(dat, aes(x=category, y=log10p, color=category)) + 
    geom_point(size=2, position = "jitter") +
 geom_text(aes(label=ifelse(label=="yes",trait,'')),hjust=.5,vjust=0, angle=0,nudge_y = 1)+
  xlab("External Trait Category") +
  ylab("-log10 p-value") +
    ylim(c(0,350))+
    ggtitle(i)+
  geom_hline(yintercept = 5.301, linetype="dashed")+
  theme(legend.title = element_blank(),
        panel.border = element_blank(),
      panel.grid.major = element_blank(),
          axis.ticks.x=element_blank(),
      axis.text.x=element_blank(),
      panel.grid.minor = element_blank())
  
  
  phewas_plt_list[[i]]<- plt
}

phewas_plt<- ggarrange(phewas_plt_list[[2]], phewas_plt_list[[1]], phewas_plt_list[[3]], nrow=3, common.legend = T)

ggsave(plot = phewas_plt,filename = "/Users/claire/Desktop/sleepGWAS/figs/PHEWAS_Full.PNG", width = 8, height = 10, units = "in")

```

## phewas for single indicator factors

```{r}
insom<- fread("/Users/claire/Desktop/sleepGWAS/FUMA_out/insom/GenomicRiskLoci.txt", header=T, data.table=F)
head(insom); nrow(insom)

insom<- insom$rsID

dur<- fread("/Users/claire/Desktop/sleepGWAS/FUMA_out/duration/GenomicRiskLoci.txt", header=T, data.table=F)
head(dur); nrow(dur)

dur<- dur$rsID


insom1<- insom[1:22]
insom2<- insom[23:length(insom)]

insomPHE1<- phewas(insom1, pval = 1e-05, batch = c(), access_token = check_access_token())
insomPHE2<- phewas(insom2, pval = 1e-05, batch = c(), access_token = check_access_token())

insomPHE<- rbind(insomPHE1, insomPHE2)

insomPHE$factor<- "Non-Insomnia"

#fwrite(insomPHE, "/Users/claire/Desktop/sleepGWAS/PHEWASinsomnia.csv", sep='\t')

dur1<- dur[1:22]
dur2<- dur[23:44]
dur3<- dur[45:65]
dur4<- dur[66:length(dur)]

durPHE1<- phewas(dur1, pval = 1e-05, batch = c(), access_token = check_access_token())
durPHE2<- phewas(dur2, pval = 1e-05, batch = c(), access_token = check_access_token())
durPHE3<- phewas(dur3, pval = 1e-05, batch = c(), access_token = check_access_token())
durPHE4<- phewas(dur4, pval = 1e-05, batch = c(), access_token = check_access_token())

durPHE<- rbind(durPHE1, durPHE2, durPHE3, durPHE4)

durPHE$factor<- rep("Duration", 1, nrow(durPHE))

#fwrite(durPHE, "/Users/claire/Desktop/sleepGWAS/PHEWASduration.csv", sep='\t')
```

### non insom PHEWAS
```{r}

### might want to do the below command before subsetting to get rid of the unclassified traits in the phewas file. this way, any that are unclassified will carry over

# insomPHE$category <- factor(insomPHE$trait, levels=phewas$trait, labels=phewas$category)

### ^ need to write and classify additional unclassified

#fwrite(insomPHE, "/Users/claire/Desktop/sleepGWAS/PHEWASinsomnia.csv", sep='\t')

insomPHE<- fread("/Users/claire/Desktop/sleepGWAS/PHEWASinsomnia.csv", header=T, data.table=F)

insomPHE$log10p<- -log10(insomPHE$p)
insomPHE$log10p[insomPHE$log10p==Inf]<- 10

insomPHE$trait<- sub(".*Non-cancer illness code self-reported: malabsorption/", "", insomPHE$trait)


insomPHE<- insomPHE %>%
  filter(category!="Unclassified")

Insom_labels<- c("Blood Assay", "Health", "Reproductive", "Sleep", "Cardiovascular", "Anthropometric", "Education", "Lifecourse")

insomPHE$trait <- str_wrap(insomPHE$trait, width = 10)


# insom
insom_labeled<- data.frame()
insom_dat<- list()
for (c in unique(insomPHE$category)) {
  insom_dat[[c]]<- insomPHE %>% filter(category==c) %>%
    arrange(desc(log10p))
  insom_dat[[c]]$label<- ifelse(rownames(insom_dat[[c]])==1 & insom_dat[[c]]$category %in% Insom_labels, "yes", NA)
  insom_labeled<- rbind(insom_dat[[c]], insom_labeled)
}


plt_insom<- ggplot(insom_labeled, aes(x=category, y=log10p, color=category)) + 
    geom_point(size=2, position = "jitter") +
 geom_text(aes(label=ifelse(label=="yes",trait,'')),hjust=.5,vjust=0, angle=0,nudge_y = 1)+
  xlab("External Trait Category") +
  ylab("-log10 p-value") +
    ylim(c(0,350))+
    ggtitle("Non-Insomnia PheWAS")+
  geom_hline(yintercept = 5.301, linetype="dashed")+
  theme(legend.title = element_blank(),
        panel.border = element_blank(),
      panel.grid.major = element_blank(),
          axis.ticks.x=element_blank(),
      axis.text.x=element_blank(),
      panel.grid.minor = element_blank())





```

### sleep duration PHEWAS
```{r}

### might want to do the below command before subsetting to get rid of the unclassified traits in the phewas file. this way, any that are unclassified will carry over

#durPHE$category <- factor(durPHE$trait, levels=phewas$trait, labels=phewas$category)

### ^ need to write and classify additional unclassified

#fwrite(durPHE, "/Users/claire/Desktop/sleepGWAS/PHEWASduration.csv", sep='\t')

durPHE<- fread("/Users/claire/Desktop/sleepGWAS/PHEWASduration.csv", header=T, data.table = F)

durPHE$log10p<- -log10(durPHE$p)
durPHE$log10p[durPHE$log10p==Inf]<- 10

durPHE$trait<- sub(".*Non-cancer illness code self-reported: malabsorption/", "", durPHE$trait)


durPHE<- durPHE %>%
  filter(category!="Unclassified")

Dur_labels<- c("Blood Assay", "Health", "Reproductive", "Sleep", "Cardiovascular", "Anthropometric", "Cognitive", "Diet")

durPHE$trait <- str_wrap(durPHE$trait, width = 10)


# duration
dur_labeled<- data.frame()
dur_dat<- list()
for (c in unique(durPHE$category)) {
  dur_dat[[c]]<- durPHE %>% filter(category==c) %>%
    arrange(desc(log10p))
  dur_dat[[c]]$label<- ifelse(rownames(dur_dat[[c]])==1 & dur_dat[[c]]$category %in% Dur_labels, "yes", NA)
  dur_labeled<- rbind(dur_dat[[c]], dur_labeled)
}


plt_dur<- ggplot(dur_labeled, aes(x=category, y=log10p, color=category)) + 
    geom_point(size=2, position = "jitter") +
 geom_text(aes(label=ifelse(label=="yes",trait,'')),hjust=.5,vjust=0, angle=0,nudge_y = 1)+
  xlab("External Trait Category") +
  ylab("-log10 p-value") +
    ylim(c(0,350))+
    ggtitle("Sleep Duration PheWAS")+
  geom_hline(yintercept = 5.301, linetype="dashed")+
  theme(legend.title = element_blank(),
        panel.border = element_blank(),
      panel.grid.major = element_blank(),
          axis.ticks.x=element_blank(),
      axis.text.x=element_blank(),
      panel.grid.minor = element_blank())

### arrange and save single indicator PHEWAS

phewas_plts_single<- ggarrange(plt_dur, plt_insom, nrow=2, common.legend = T)

ggsave(plot = phewas_plts_single,filename = "/Users/claire/Desktop/sleepGWAS/figs/PHEWAS_single_indicators.PNG", width = 7, height = 10, units = "in")



for (c in unique(durPHE$category)) {
  print(durPHE %>% filter(category==c) %>%
          select(trait, category, p, beta) %>%
          arrange(-desc(p)) %>%
          slice(1:3))
  }


for (c in unique(insomPHE$category)) {
  print(insomPHE %>% filter(category==c) %>%
          select(trait, category, p, beta) %>%
          arrange(-desc(p)) %>%
          slice(1:3))
  }



```

word cloud lol

```{r}
library("htmlwidgets")
library(wordcloud2) 

## alert
alert<- phewas %>% filter(factor=="Alertness") %>%
  select(category) %>%
  count(category) %>%
  distinct(category, n) 

alert<- wordcloud2( alert, color='random-light' , backgroundColor="#00BFC4")

saveWidget(alert,"/Users/claire/Desktop/tmp.html",selfcontained = F)
webshot("/Users/claire/Desktop/tmp.html","/Users/claire/Desktop/sleepGWAS/figs/alertness_WC.pdf", delay =5, vwidth = 480, vheight=480)

## circ
circ<- phewas %>% filter(factor=="Circadian Preference") %>%
  select(category) %>%
  count(category) %>%
  distinct(category, n) 

circ<- wordcloud2( circ, color='random-light' , backgroundColor="brown2")

saveWidget(circ,"/Users/claire/Desktop/tmp.html",selfcontained = F)
webshot("/Users/claire/Desktop/tmp.html","/Users/claire/Desktop/sleepGWAS/figs/circ_WC.pdf", delay =5, vwidth = 480, vheight=480)

## effic
effic<- phewas %>% filter(factor=="Efficiency") %>%
  select(category) %>%
  count(category) %>%
  distinct(category, n) 

effic<- wordcloud2( effic, color='random-light' , backgroundColor="#619CFF")

saveWidget(effic,"/Users/claire/Desktop/tmp.html",selfcontained = F)
webshot("/Users/claire/Desktop/tmp.html","/Users/claire/Desktop/sleepGWAS/figs/efficiency_WC.pdf", delay =5, vwidth = 480, vheight=480)
```


## do a phewas for SNPs that were in the consituent but not factor GWAS

```{r}

circ<- fread("/Users/claire/Desktop/Desktop/sleepGWAS/circ_SNPs_not_in_factor.txt", header=T, data.table=F)

nrow(circ) # 167

alert<- fread("/Users/claire/Desktop/Desktop/sleepGWAS/alert_SNPs_not_in_factor.txt", header=T, data.table = F)

nrow(alert) # 126

effic<- fread("/Users/claire/Desktop/Desktop/sleepGWAS/eff_SNPs_not_in_factor.txt", header=T, data.table = F)

nrow(effic) # 23

```

circ component phewas not in our factor

```{r}
#circ<- circ$SNP

#circ1<- circ[1:30]
# circ2<- circ[31:60]
# circ3<- circ[61:90]
# circ4<- circ[91:120]
# circ5<- circ[121:150]
# circ6<- circ[151:167]
# 
# 
# #circPHE1<- phewas(circ1, pval = 1e-05, batch = c(), access_token = check_access_token())
# circPHE2<- phewas(circ2, pval = 1e-05, batch = c(), access_token = check_access_token())
# circPHE3<- phewas(circ3, pval = 1e-05, batch = c(), access_token = check_access_token())
# circPHE4<- phewas(circ4, pval = 1e-05, batch = c(), access_token = check_access_token())
# circPHE5<- phewas(circ5, pval = 1e-05, batch = c(), access_token = check_access_token())
# circPHE6<- phewas(circ6, pval = 1e-05, batch = c(), access_token = check_access_token())
# 
# circPHE<- rbind(circPHE1, circPHE2, circPHE3, circPHE4, circPHE5, circPHE6)
# 
# length(unique(circPHE$rsid))
# 
# 
# fwrite(circPHE, "/Users/claire/Desktop/circ_phe_tmp.txt", sep='\t') ## write just in case we crash, delete this after combined

```


alert component phewas not in our factor

```{r}
# alert<- alert$SNP
# 
# alert1<- alert[1:30]
# alert2<- alert[31:60]
# alert3<- alert[61:90]
# alert4<- alert[91:110]
# alert5<- alert[111:126]
# 
# alertPHE1<- phewas(alert1, pval = 1e-05, batch = c(), access_token = check_access_token())
# alertPHE2<- phewas(alert2, pval = 1e-05, batch = c(), access_token = check_access_token())
# alertPHE3<- phewas(alert3, pval = 1e-05, batch = c(), access_token = check_access_token())
# alertPHE4<- phewas(alert4, pval = 1e-05, batch = c(), access_token = check_access_token())
# alertPHE5<- phewas(alert5, pval = 1e-05, batch = c(), access_token = check_access_token())


#alertPHE<- rbind(alertPHE1, alertPHE2, alertPHE3,alertPHE4, alertPHE5)

#length(unique(alertPHE$rsid))

#fwrite(alertPHE, "/Users/claire/Desktop/alert_phe_tmp.txt", sep='\t') ## write just in case we crash, delete this after combined


```

effic component phewas not in our factor

```{r}
#effic<- effic$SNP
# 
# efficPHE<- phewas(effic, pval = 1e-05, batch = c(), access_token = check_access_token())
# 
# head(efficPHE); nrow(efficPHE)
# 
# length(unique(efficPHE$rsid))
# 
# 
# fwrite(efficPHE, "/Users/claire/Desktop/effic_phe_tmp.txt", sep='\t') ## write just in case we crash, delete this after combined
```

take full exclusive phewas and just see what is also in the other phewas...


```{r}
#full_ex_phe<- rbind(circPHE, alertPHE, efficPHE)
#nrow(full_ex_phe) ### 1

#head(full_ex_phe)

#fwrite(full_ex_phe, "/Users/claire/Desktop/Desktop/sleepGWAS/non-factor_PHEWAS.csv", sep='\t')
```

10693 external associations

these don't really seem to be that different. the associations that did not overlap were ones that i did not categorize... like Spherical power (left) or SHBG or ENSG00000258429

```{r}
#phewas<- fread("/Users/claire/Desktop/Desktop/sleepGWAS/PHEWAS.csv", header=T, data.table=F)

phewas<- fread("/Users/claire/Desktop/sleepGWAS/PHEWAS.csv", header=T, data.table=F)


head(phewas); nrow(phewas)

full_ex_phe<- fread("/Users/claire/Desktop//sleepGWAS/phewas_component.csv", header=T, data.table=F)

table(full_ex_phe$factor); table(phewas$factor)


# circ
circ<- phewas %>% filter(factor=="Circadian Preference")
exep_circ<- full_ex_phe %>% filter(factor=="circ_comp") 

nrow(circ); nrow(exep_circ) # 2663 from factor  vs 5130 from indicators

length(unique(exep_circ$trait)); length(unique(circ$trait))

sum(circ$trait %in% exep_circ$trait) # 2371 from factor were also in the indicators (89%)
sum(!circ$trait %in% exep_circ$trait) # 292 from factor were not indicators which adds up

factor_not_inds<- circ[!circ$trait %in% exep_circ$trait, c(12,14)]
subset(factor_not_inds, factor_not_inds$category!="Unclassified")
table(factor_not_inds$category)

### factor traits not in indicator mostly cardiovascular, health etc nothing special

inds_not_factor<- exep_circ[!exep_circ$trait %in% circ$trait, c(12,14)]
subset(inds_not_factor, inds_not_factor$category!="Unclassified")
table(inds_not_factor$category)


# alert
alert<- phewas %>% filter(factor=="Alertness")
exep_alert<- full_ex_phe %>% filter(factor=="alert_comp") 

nrow(alert); nrow(exep_alert) # 1757 from factor  vs 3732 from indicators

sum(alert$trait %in% exep_alert$trait) # 1561 from factor were also in the indicators (88%)
sum(!alert$trait %in% exep_alert$trait) # 196 from factor were not indicators which adds up

factor_not_inds<- alert[!alert$trait %in% exep_alert$trait, c(12,14)]
subset(factor_not_inds, factor_not_inds$category!="Unclassified")
table(factor_not_inds$category)

### factor traits not in indicator mostly cardiovascular, health etc nothing special

inds_not_factor<- exep_alert[!exep_alert$trait %in% alert$trait, c(12,14)]
subset(inds_not_factor, inds_not_factor$category!="Unclassified")
table(inds_not_factor$category)


# effic
effic<- phewas %>% filter(factor=="Efficiency")
exep_effic<- full_ex_phe %>% filter(factor=="effic_comp") 

nrow(effic); nrow(exep_effic) # 151 from factor  vs 1424 from indicators

sum(effic$trait %in% exep_effic$trait) # 137 from factor were also in the indicators (91%)
sum(!effic$trait %in% exep_effic$trait) # 14 from factor were not indicators which adds up

factor_not_inds<- effic[!effic$trait %in% exep_effic$trait, c(12,14)]
subset(factor_not_inds, factor_not_inds$category!="Unclassified")
table(factor_not_inds$category)

### factor traits not in indicator mostly cardiovascular, health etc nothing special

inds_not_factor<- exep_effic[!exep_effic$trait %in% effic$trait, c(12,14)]
subset(inds_not_factor, inds_not_factor$category!="Unclassified")
table(inds_not_factor$category)



### bigger scale
overlap<- intersect(phewas$trait, full_ex_phe$trait)
length(overlap)

overlap<- data.frame(overlap)
colnames(overlap)[1]<- 'trait'

phewas %>% select(trait, category) %>%
  right_join(overlap) %>%
  group_by(category) %>%
  count(trait) %>%
  slice(1) %>%
 ggplot(aes(y=n, x=category, fill=category)) +
  geom_col()


```

### plot phewasv.2 to visualize

```{r}

#phe_cat<- phewas %>% select(trait, category)

#full_ex_phe$category <- factor(full_ex_phe$trait, levels=phe_cat$trait, labels=phe_cat$category)

#full_ex_phe$category<- as.character(full_ex_phe$category)


full_ex_phe<- fread("/Users/claire/Desktop/sleepGWAS/phewas_component.csv", header=T, data.table=F)

full_ex_phe$log10p<- -log10(full_ex_phe$p)
full_ex_phe$log10p[full_ex_phe$log10p==Inf]<- 10

full_ex_phe$trait<- sub(".*Non-cancer illness code self-reported: malabsorption/", "", full_ex_phe$trait)


full_ex_phe<- full_ex_phe %>%
  filter(category!="Unclassified")

Alertness_labels<- c("Blood Assay", "Health", "Reproductive", "Sleep", "Cardiovascular", "Anthropometric", "Cognitive")
Circadian_Preference_labels<- c("Anthropometric", "Health", "Reproductive", "Cardiovascular", "Blood Assay", "Education", "Recreation", "Sleep", "Lifecourse", "Anthropometric", "Puberty")
Efficiency_labels<- c("Blood Assay", "Cardiovascular", "Diet", "Lifecourse", "Psychiatric", "Sleep", "Health", "Reproductive")

full_ex_phe$trait <- str_wrap(full_ex_phe$trait, width = 10)


# effic
effic_labeled<- data.frame()
effic_dat<- list()
effic<- full_ex_phe %>% filter(factor=="effic_comp")  %>%
  mutate(factor="Efficiency indicator loci not in factor")
for (c in unique(effic$category)) {
  effic_dat[[c]]<- effic %>% filter(category==c) %>%
    arrange(desc(log10p))
  effic_dat[[c]]$label<- ifelse(rownames(effic_dat[[c]])==1 & effic_dat[[c]]$category %in% Efficiency_labels, "yes", NA)
  effic_labeled<- rbind(effic_dat[[c]], effic_labeled)
}

# circ 
circ_labeled<- data.frame()
circ_dat<- list()
circ<- full_ex_phe %>% filter(factor=="circ_comp")  %>%
  mutate(factor="Ciracdian Preference indicator loci not in factor")
for (c in unique(circ$category)) {
  circ_dat[[c]]<- circ %>% filter(category==c) %>%
    arrange(desc(log10p))
  circ_dat[[c]]$label<- ifelse(rownames(circ_dat[[c]])==1 & circ_dat[[c]]$category %in% Circadian_Preference_labels, "yes", NA)
  circ_labeled<- rbind(circ_dat[[c]], circ_labeled)
}

# alert
alert_labeled<- data.frame()
alert_dat<- list()
alert<- full_ex_phe %>% filter(factor=="alert_comp")  %>%
  mutate(factor="Alertness indicator loci not in factor")
for (c in unique(alert$category)) {
  alert_dat[[c]]<- alert %>% filter(category==c) %>%
    arrange(desc(log10p))
  alert_dat[[c]]$label<- ifelse(rownames(alert_dat[[c]])==1 & alert_dat[[c]]$category %in% Alertness_labels, "yes", NA)
  alert_labeled<- rbind(alert_dat[[c]], alert_labeled)
}


nrow(alert_labeled); nrow(circ_labeled); nrow(effic_labeled)
table(phewas$factor) ## checks out

phefull<- rbind(alert_labeled, circ_labeled, effic_labeled)


phewas_plt_list<- list()
for (i in unique(phefull$factor)) {
  dat<- phefull %>% filter(factor==i)
  plt<- ggplot(dat, aes(x=category, y=log10p, color=category)) + 
    geom_point(size=2, position = "jitter") +
 geom_text(aes(label=ifelse(label=="yes",trait,'')),hjust=.5,vjust=0, angle=0,nudge_y = 1)+
  xlab("External Trait Category") +
  ylab("-log10 p-value") +
    ylim(c(0,350))+
    ggtitle(i)+
  geom_hline(yintercept = 5.301, linetype="dashed")+
  theme(legend.title = element_blank(),
        panel.border = element_blank(),
      panel.grid.major = element_blank(),
          axis.ticks.x=element_blank(),
      axis.text.x=element_blank(),
      panel.grid.minor = element_blank())
  
  
  phewas_plt_list[[i]]<- plt
}


phewas_plt2<- ggarrange(phewas_plt_list[[2]], phewas_plt_list[[1]], phewas_plt_list[[3]], nrow=3, common.legend = T)

ggsave(plot = phewas_plt2,filename = "/Users/claire/Desktop/sleepGWAS/figs/PHEWAS_from_indicators.PNG", width = 9, height = 11, units = "in")


c<- subset(full_ex_phe, full_ex_phe$factor=="circ_comp")
nrow(c) # should be 5344

a<- subset(full_ex_phe, full_ex_phe$factor=="alert_comp")
nrow(a) # should be 3860

e<- subset(full_ex_phe, full_ex_phe$factor=="effic_comp")
nrow(e) # should be 1489

#fwrite(full_ex_phe, "/Users/claire/Desktop/phewas_component.csv", sep='\t')

```
# save all PHEWAS' into table for supplement

needs to be such that each category has their own column ... organized by p value

each phewas will be it's own tab/ table

### factor phewas tables

```{r}
phewas<- fread("/Users/claire/Desktop/sleepGWAS/PHEWAS.csv", header=T, data.table=F)
head(phewas)

table(phewas$factor)

dat<- list()
for (i in unique(phewas$factor)) {
  dat[[i]]<- phewas %>% filter(factor==i) %>%
    select(factor, category, chr, rsid, ea, nea, eaf, beta, p, trait, n)
  for (c in unique(dat[[i]]$category)) {
    dat[[paste0(i,c)]]<- dat[[i]] %>% filter(category==c) %>%
      arrange(-desc(p))
    fwrite(dat[[paste0(i,c)]], paste0('/Users/claire/Desktop/', i, c, ".csv"))
  }
}


names(dat)

### no good way to cbind or merge since row numbers differ and column names are the same.... save and copy into one dataframe [:


```


### indicator phewas tables

```{r}
phewas<- fread("/Users/claire/Desktop/sleepGWAS/phewas_component.csv", header=T, data.table=F)


table(phewas$factor)

dat<- list()
for (i in unique(phewas$factor)) {
  dat[[i]]<- phewas %>% filter(factor==i) %>%
    select(factor, category, chr, rsid, ea, nea, eaf, beta, p, trait, n)
  for (c in unique(dat[[i]]$category)) {
    dat[[paste0(i,c)]]<- dat[[i]] %>% filter(category==c) %>%
      arrange(-desc(p))
    fwrite(dat[[paste0(i,c)]], paste0('/Users/claire/Desktop/', i, c, ".csv"))
  }
}


names(dat)

```


### single var phewas tables

duration

```{r}
phewas<- fread("/Users/claire/Desktop/sleepGWAS/PHEWASduration.csv", header=T, data.table=F)

table(phewas$factor)

dat<- list()
for (i in unique(phewas$factor)) {
  dat[[i]]<- phewas %>% filter(factor==i) %>%
    select(factor, category, chr, rsid, ea, nea, eaf, beta, p, trait, n)
  for (c in unique(dat[[i]]$category)) {
    dat[[paste0(i,c)]]<- dat[[i]] %>% filter(category==c) %>%
      arrange(-desc(p))
    fwrite(dat[[paste0(i,c)]], paste0('/Users/claire/Desktop/', "Duration", c, ".csv"))
  }
}


names(dat)
```

Insomnia

```{r}
phewas<- fread("/Users/claire/Desktop/sleepGWAS/PHEWASinsomnia.csv", header=T, data.table=F)

table(phewas$factor)

dat<- list()
for (i in unique(phewas$factor)) {
  dat[[i]]<- phewas %>% filter(factor==i) %>%
    select(factor, category, chr, rsid, ea, nea, eaf, beta, p, trait, n)
  for (c in unique(dat[[i]]$category)) {
    dat[[paste0(i,c)]]<- dat[[i]] %>% filter(category==c) %>%
      arrange(-desc(p))
    fwrite(dat[[paste0(i,c)]], paste0('/Users/claire/Desktop/', "Insomnia", c, ".csv"))
  }
}


names(dat)
```

